<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>Preparing the input - immunopipe</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
        <link href="../assets/style.css" rel="stylesheet" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Preparing the input";
        var mkdocs_page_input_path = "preparing-input.md";
        var mkdocs_page_url = null;
      </script>
    
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.configure({ ignoreUnescapedHTML: true }); hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="..">
          <img src="../rtd-logo.png" class="logo" alt="Logo"/>
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="nav-scrollable-wrapper">
      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../introduction/">Introduction</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../installation/">Installation</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../getting-started/">Geting started</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../gallery/">Gallery</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../running/">Running the pipeline</a>
                </li>
              </ul>
              <ul class="current">
                <li class="toctree-l1 current"><a class="reference internal current" href="./">Preparing the input</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#single-cell-rna-seq-scrna-seq-data">Single-cell RNA-seq (scRNA-seq) data</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#single-cell-tcr-seq-sctcr-seq-data">Single-cell TCR-seq (scTCR-seq) data</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#metadata">Metadata</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#other-optional-files">Other optional files</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#genesfeatures-to-visualize-for-seurat-object">Genes/Features to visualize for Seurat object</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#pathways-for-gene-set-enrichment-analysis-gsea">Pathways for Gene Set Enrichment Analysis (GSEA)</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#cell-type-database-for-cell-type-annotation-by-sctype-or-hitype">Cell type database for cell type annotation by sctype or hitype</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#model-for-cell-type-annotation-by-celltypist">Model for cell type annotation by celltypist</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#metabolic-pathway-for-metabolic-landscape-analysis">Metabolic pathway for Metabolic Landscape Analysis</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#reference-for-seurat-mapping-if-you-want-to-perform-supervised-clustering">Reference for Seurat mapping if you want to perform supervised clustering</a>
    </li>
        </ul>
    </li>
    </ul>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../configurations/">Configurations</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">Processes</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/SampleInfo/">SampleInfo</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/ImmunarchLoading/">ImmunarchLoading</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/SeuratPreparing/">SeuratPreparing</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/SeuratClusteringOfAllCells/">SeuratClusteringOfAllCells</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/ClusterMarkersOfAllCells/">ClusterMarkersOfAllCells</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/TopExpressingGenesOfAllCells/">TopExpressingGenesOfAllCells</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/TCellSelection/">TCellSelection</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/ModuleScoreCalculator/">ModuleScoreCalculator</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/SeuratClustering/">SeuratClustering</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/SeuratMap2Ref/">SeuratMap2Ref</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/CellTypeAnnotation/">CellTypeAnnotation</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/SeuratSubClustering/">SeuratSubClustering</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/TCRClustering/">TCRClustering</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/TCRClusterStats/">TCRClusterStats</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/IntegratingTCRClusters/">IntegratingTCRClusters</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/ClusterMarkers/">ClusterMarkers</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/TopExpressingGenes/">TopExpressingGenes</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/Immunarch/">Immunarch</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/TESSA/">TESSA</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/IntegratingTCR/">IntegratingTCR</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/SeuratClusterStats/">SeuratClusterStats</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/RadarPlots/">RadarPlots</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/CloneResidency/">CloneResidency</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/ScFGSEA/">ScFGSEA</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/CellsDistribution/">CellsDistribution</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/MarkersFinder/">MarkersFinder</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/MetaMarkers/">MetaMarkers</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/CDR3AAPhyschem/">CDR3AAPhyschem</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/CellCellCommunication/">CellCellCommunication</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/CellCellCommunicationPlots/">CellCellCommunicationPlots</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Metabolic Landscape Analysis</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/ScrnaMetabolicLandscape/">Introduction & Group Arguments</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/MetabolicInput/">MetabolicInput</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/MetabolicExprImpution/">MetabolicExprImpution</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/MetabolicFeatures/">MetabolicFeatures</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/MetabolicFeaturesIntraSubset/">MetabolicFeaturesIntraSubset</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/MetabolicPathwayActivity/">MetabolicPathwayActivity</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/MetabolicPathwayHeterogeneity/">MetabolicPathwayHeterogeneity</a>
                  </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../faq/">FAQ</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../CHANGELOG/">Change log</a>
                </li>
              </ul>
      </div>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="..">immunopipe</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".." class="icon icon-home" aria-label="Docs"></a></li>
      <li class="breadcrumb-item active">Preparing the input</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/pwwang/immunopipe/blob/dev/docs/preparing-input.md">Edit on pwwang/immunopipe</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h2 id="preparing-the-input-data">Preparing the input data<a class="headerlink" href="#preparing-the-input-data" title="Permanent link">&para;</a></h2>
<h3 id="single-cell-rna-seq-scrna-seq-data">Single-cell RNA-seq (scRNA-seq) data<a class="headerlink" href="#single-cell-rna-seq-scrna-seq-data" title="Permanent link">&para;</a></h3>
<p>Currently only <a href="https://www.10xgenomics.com/">10X Genomics</a> data are supported.</p>
<p>For each sample, you need to provide a directory containing the files that are generated by <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger"><code>CellRanger</code></a>. Specifically, the directory should be able to be read by <a href="https://satijalab.org/seurat/reference/read10x"><code>Seurat::Read_10X()</code></a>. For example, the directory should contain <code>matrix.mtx</code>, <code>barcodes.tsv</code> and <code>features.tsv</code>.
These files can also be gzipped.</p>
<p>You can also use the <code>h5</code> file generated by <code>CellRanger</code>.</p>
<h3 id="single-cell-tcr-seq-sctcr-seq-data">Single-cell TCR-seq (scTCR-seq) data<a class="headerlink" href="#single-cell-tcr-seq-sctcr-seq-data" title="Permanent link">&para;</a></h3>
<p>The scRNA-seq data is optional for the pipeline. However, the scTCR-seq data is required for the pipeline.</p>
<p>The scTCR-seq data, if available, should be paired with the scRNA-seq data. Theoratically, as long as the data can be loaded by <a href="https://immunarch.com/reference/repLoad.html"><code>immunarch::repLoad()</code></a>, it should be fine. However, the pipeline is only tested with the data generated by <a href="https://support.10xgenomics.com/single-cell-vdj/software/pipelines/latest/output/annotation"><code>CellRanger</code></a>. Specifically, the directory should contain <code>filtered_contig_annotations.csv</code> file, or at least the <code>all_contig_annotations.csv</code> file.</p>
<h3 id="metadata">Metadata<a class="headerlink" href="#metadata" title="Permanent link">&para;</a></h3>
<p>A metadata file is required as an input file for the pipeline. It should be a <code>TAB</code> delimited file with 3 required columns:</p>
<ul>
<li><code>Sample</code>: A unique id for each sample</li>
<li><code>RNAData</code>: The directory or <code>h5</code> file for single-cell RNA data for this sample, as described above.</li>
<li><code>TCRData</code> (optional): The directory for single-cell TCR data for this sample as described above.</li>
</ul>
<p>When <code>TCRData</code> is not provided, the pipeline will skip the processes related to scTCR-seq data (see <a href="../introduction/#only-scrna-seq-data-avaiable">Routes of the pipeline</a> for more details).
You can also add other columns to the metadata file. The columns will be added to both:</p>
<ul>
<li>meta data of the object loaded by <code>immunarch::repLoad()</code> (i.e. <code>data$meta</code>)</li>
<li>meta data of the seurat object loaded by <code>Seurat::Read10X()</code> or <code>Seurat::Read10X_h5()</code> (i.e. <code>srtobj@meta.data</code>)</li>
</ul>
<p>This file should be provided to <code>SampleInfo</code> process. See <a href="../processes/SampleInfo/"><code>SampleInfo</code></a> for more details.</p>
<p>An example metadata file can be found <a href="https://github.com/pwwang/immunopipe-example/blob/master/data/samples.txt">here</a>.</p>
<p>You can also use <a href="../processes/SampleInfo/"><code>SampleInfo</code></a> with <code>envs.save_mutated = true</code> and/or <a href="../processes/IntegratingTCR/"><code>IntegratingTCR</code></a> to add columns to metadata by configuration. These columns are persisted for downstream analysis. The difference is that <code>SampleInfo</code> can only use the columns that are already in the metadata file, while <code>IntegratingTCR</code> can use the columns that are generated by the pipeline (e.g. TCR clone information).</p>
<h3 id="other-optional-files">Other optional files<a class="headerlink" href="#other-optional-files" title="Permanent link">&para;</a></h3>
<h4 id="genesfeatures-to-visualize-for-seurat-object">Genes/Features to visualize for Seurat object<a class="headerlink" href="#genesfeatures-to-visualize-for-seurat-object" title="Permanent link">&para;</a></h4>
<p>If you have a set of genes/features of interest, you can provide a file with those genes, one gene per line, to <code>SeuratClusterStats.envs.exprs.features</code> for:</p>
<ul>
<li>Ridge plots using <a href="https://satijalab.org/seurat/reference/ridgeplot"><code>Seurat::RidgePlot()</code></a>,</li>
<li>Violin plots using <a href="https://satijalab.org/seurat/reference/vlnplot"><code>Seurat::VlnPlot()</code></a>,</li>
<li>Feature plots using <a href="https://satijalab.org/seurat/reference/featureplot"><code>Seurat::FeaturePlot()</code></a>,</li>
<li>Dot plots using <a href="https://satijalab.org/seurat/reference/dotplot"><code>Seurat::DotPlot()</code></a>, and</li>
<li>Heatmaps using <a href="https://satijalab.org/seurat/reference/doheatmap"><code>Seurat::DoHeatmap()</code></a>.</li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The genes should exist in the RNA-seq data (i.e <code>features.tsv</code> or the <code>h5</code> file from cellranger).</p>
</div>
<p>See <a href="../processes/SeuratClusterStats/"><code>SeuratClusterStats</code></a> for more details.</p>
<h4 id="pathways-for-gene-set-enrichment-analysis-gsea">Pathways for Gene Set Enrichment Analysis (GSEA)<a class="headerlink" href="#pathways-for-gene-set-enrichment-analysis-gsea" title="Permanent link">&para;</a></h4>
<p>If you want to perform GSEA, you need to provide a file containing the pathways. The file should be in the <a href="https://docs.gsea-msigdb.org/#GSEA/Data_Formats/#gmt-gene-matrix-transposed-file-format-gmt">GMT format</a>. You can provide the file to <code>ScFGSEA.envs.gmtfile</code>. Similarly, the genes should exist (be in the same format) in the <code>features.tsv</code> file.</p>
<p>See <a href="../processes/ScFGSEA/"><code>ScFGSEA</code></a> for more details.</p>
<p>You can also find an example here: <a href="https://github.com/pwwang/immunopipe-example/blob/master/data/MSigDB_Hallmark_v7.5.1.gmt">https://github.com/pwwang/immunopipe-example/blob/master/data/MSigDB_Hallmark_v7.5.1.gmt</a></p>
<h4 id="cell-type-database-for-cell-type-annotation-by-sctype-or-hitype">Cell type database for cell type annotation by <code>sctype</code> or <code>hitype</code><a class="headerlink" href="#cell-type-database-for-cell-type-annotation-by-sctype-or-hitype" title="Permanent link">&para;</a></h4>
<p>If you want to perform cell type annotation, you need to provide a file containing the cell type database if you are using <code>sctype</code> or <code>hitype</code>. The database file should be fed to <code>CellTypeAnnotation.envs.sctype_db</code> if you are using <code>sctype</code>, or <code>CellTypeAnnotation.envs.hitype_db</code> if you are using <code>hitype</code>. Again, the markers in the database should exist (be in the same format) in the <code>features.tsv</code> file or the <code>h5</code> file.</p>
<p>See <a href="../processes/CellTypeAnnotation/"><code>CellTypeAnnotation</code></a> for more details.</p>
<p>Examples can be found here: <a href="https://github.com/IanevskiAleksandr/sc-type/blob/master/ScTypeDB_short.xlsx">ScTypeDB_short.xlsx</a> and <a href="https://github.com/IanevskiAleksandr/sc-type/blob/master/ScTypeDB_full.xlsx">ScTypeDB_full.xlsx</a>.</p>
<h4 id="model-for-cell-type-annotation-by-celltypist">Model for cell type annotation by <code>celltypist</code><a class="headerlink" href="#model-for-cell-type-annotation-by-celltypist" title="Permanent link">&para;</a></h4>
<p>If you want to perform cell type annotation by <code>celltypist</code>, you need to provide a model file. The model file should be fed to <code>CellTypeAnnotation.envs.celltypist_args.model</code>. The information of models can be found <a href="https://celltypist.cog.sanger.ac.uk/models/models.json">here</a>. Download the one you want to use and provide the path to the file.</p>
<h4 id="metabolic-pathway-for-metabolic-landscape-analysis">Metabolic pathway for Metabolic Landscape Analysis<a class="headerlink" href="#metabolic-pathway-for-metabolic-landscape-analysis" title="Permanent link">&para;</a></h4>
<p>Similarly, if you want to perform metabolic landscape analysis, you need to provide a file containing the metabolic pathways. The file should be in the <a href="https://docs.gsea-msigdb.org/#GSEA/Data_Formats/#gmt-gene-matrix-transposed-file-format-gmt">GMT format</a>. You can provide the file to <code>ScMetabolicLandscape.envs.gmtfile</code>. This file can also be used for GSEA. A pathway file for KEGG metabolism is provided <a href="https://github.com/pwwang/biopipen/blob/master/tests/data/scrna_metabolic/KEGG_metabolism.gmt">here</a>.</p>
<p>See <a href="../processes/ScrnaMetabolicLandscape/"><code>ScrnaMetabolicLandscape</code></a> for more details.</p>
<h4 id="reference-for-seurat-mapping-if-you-want-to-perform-supervised-clustering">Reference for Seurat mapping if you want to perform supervised clustering<a class="headerlink" href="#reference-for-seurat-mapping-if-you-want-to-perform-supervised-clustering" title="Permanent link">&para;</a></h4>
<p>If you want to perform supervised clustering, you need to provide a reference for <a href="../processes/SeuratMap2Ref/"><code>SeuratMap2Ref</code></a>. The reference should be a <code>Seurat</code> object in <code>RDS</code> or <code>h5seurat</code> file. You can provide the reference to <code>SeuratMap2Ref.envs.ref</code>.</p>
<p>See <a href="../processes/SeuratMap2Ref/"><code>SeuratMap2Ref</code></a> for more details.</p>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../running/" class="btn btn-neutral float-left" title="Running the pipeline"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../configurations/" class="btn btn-neutral float-right" title="Configurations">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  <div class="footer-notes">
    
    Version: <div id="mike-versions"></div>
    <div class="footer-sep"></div>
    <a href="https://github.com/pwwang/immunopipe" title="Open repository in GitHub">Github Repo</a>
    <div class="footer-sep"></div>Built with <a href="https://www.mkdocs.org/">MkDocs</a> using theme <a href="https://github.com/pwwang/mkdocs-rtd">mkdocs-rtd</a>.
  </div>
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
    
      <span><a href="../running/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../configurations/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script src="../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "..";</script>
    <script src="../js/theme_extra.js"></script>
    <script src="../js/theme.js"></script>
    <script src="../js/version-select.js"></script>
      <script src="../assets/script.js"></script>
      <script src="../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
