<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>Change log - immunopipe</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
        <link href="../assets/style.css" rel="stylesheet" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Change log";
        var mkdocs_page_input_path = "CHANGELOG.md";
        var mkdocs_page_url = null;
      </script>
    
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.configure({ ignoreUnescapedHTML: true }); hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="..">
          <img src="../rtd-logo.png" class="logo" alt="Logo"/>
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="nav-scrollable-wrapper">
      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../introduction/">Introduction</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../installation/">Installation</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../getting-started/">Geting started</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../gallery/">Gallery</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../running/">Running the pipeline</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../preparing-input/">Preparing the input</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../configurations/">Configurations</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../mcp-server/">MCP Sever</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">Processes</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/SampleInfo/">SampleInfo</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/SeuratPreparing/">SeuratPreparing</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/SeuratClusteringOfAllCells/">SeuratClusteringOfAllCells</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/ClusterMarkersOfAllCells/">ClusterMarkersOfAllCells</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/TopExpressingGenesOfAllCells/">TopExpressingGenesOfAllCells</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/TOrBCellSelection/">TOrBCellSelection</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/ModuleScoreCalculator/">ModuleScoreCalculator</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/SeuratClustering/">SeuratClustering</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/CellTypeAnnotation/">CellTypeAnnotation</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/SeuratMap2Ref/">SeuratMap2Ref</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/ClusterMarkers/">ClusterMarkers</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/TopExpressingGenes/">TopExpressingGenes</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/SeuratSubClustering/">SeuratSubClustering</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/ScRepLoading/">ScRepLoading</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/ScRepCombiningExpression/">ScRepCombiningExpression</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/TCRClustering/">TCRClustering</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/TESSA/">TESSA</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/SeuratClusterStats/">SeuratClusterStats</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/ScFGSEA/">ScFGSEA</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/MarkersFinder/">MarkersFinder</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/PseudoBulkDEG/">PseudoBulkDEG</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/ClonalStats/">ClonalStats</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/CDR3AAPhyschem/">CDR3AAPhyschem</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/CellCellCommunication/">CellCellCommunication</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/CellCellCommunicationPlots/">CellCellCommunicationPlots</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Metabolic Landscape Analysis</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/ScrnaMetabolicLandscape/">Introduction & Group Arguments</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/MetabolicInput/">MetabolicInput</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/MetabolicExprImputation/">MetabolicExprImputation</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/MetabolicFeatures/">MetabolicFeatures</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/MetabolicPathwayActivity/">MetabolicPathwayActivity</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../processes/MetabolicPathwayHeterogeneity/">MetabolicPathwayHeterogeneity</a>
                  </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../faq/">FAQ</a>
                </li>
              </ul>
              <ul class="current">
                <li class="toctree-l1 current"><a class="reference internal current" href="./">Change log</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#203">2.0.3</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#202">2.0.2</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#201">2.0.1</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#200-cloud-support-enhanced-visualization-new-analysesfeatures-and-more">2.0.0 (Cloud support, enhanced visualization, new analyses/features, and more ...)</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#cloud-support">Cloud support</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#enhanced-visualization">Enhanced visualization</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#enhanced-performance">Enhanced performance</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#new-analysesfeatures">New analyses/features</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#house-keeping">House keeping</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#144">1.4.4</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#143">1.4.3</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#142">1.4.2</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#141">1.4.1</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#140">1.4.0</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#139">1.3.9</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#138">1.3.8</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#137">1.3.7</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#136">1.3.6</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#135">1.3.5</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#134-post">1.3.4-post</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#134">1.3.4</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#133">1.3.3</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#132">1.3.2</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#131">1.3.1</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#130">1.3.0</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#120">1.2.0</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#111">1.1.1</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#110">1.1.0</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#105">1.0.5</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#104">1.0.4</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#103">1.0.3</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#102">1.0.2</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#101">1.0.1</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#100">1.0.0</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#highlights">Highlights</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#breaking-changes">Breaking changes</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#changes">Changes</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#features">Features</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#dependencies">Dependencies</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#0112">0.11.2</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#0111">0.11.1</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#0110">0.11.0</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#0101">0.10.1</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#0100">0.10.0</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#093">0.9.3</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#092">0.9.2</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#091">0.9.1</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#090">0.9.0</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#housekeeping-and-docs">Housekeeping and docs</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#fixedenhanced">Fixed/Enhanced</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#added">Added</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#breaking-changes-090">Breaking changes-0.9.0</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#083">0.8.3</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#082">0.8.2</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#081">0.8.1</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#080">0.8.0</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#housekeeping-and-docs-updates">Housekeeping and docs updates</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#added-080">Added-0.8.0</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#fixed">Fixed</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#breaking-changes-080">Breaking changes-0.8.0</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#070">0.7.0</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#housekeeping-and-docs-updates-070">Housekeeping and docs updates-0.7.0</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#added-070">Added-0.7.0</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#fixed-070">Fixed-0.7.0</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#breaking-changes-070">Breaking changes-0.7.0</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#060">0.6.0</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#051">0.5.1</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#050">0.5.0</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#040">0.4.0</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#030">0.3.0</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#024">0.2.4</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#023">0.2.3</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#022">0.2.2</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#021">0.2.1</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#020">0.2.0</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#011">0.1.1</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#010">0.1.0</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#007">0.0.7</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#006">0.0.6</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#005">0.0.5</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#004">0.0.4</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#003">0.0.3</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#002">0.0.2</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#001">0.0.1</a>
    </li>
    </ul>
                </li>
              </ul>
      </div>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="..">immunopipe</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".." class="icon icon-home" aria-label="Docs"></a></li>
      <li class="breadcrumb-item active">Change log</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/pwwang/immunopipe/blob/dev/docs/CHANGELOG.md">Edit on pwwang/immunopipe</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h2 id="change-log">Change Log<a class="headerlink" href="#change-log" title="Permanent link">&para;</a></h2>
<h3 id="203">2.0.3<a class="headerlink" href="#203" title="Permanent link">&para;</a></h3>
<ul>
<li>feat: adopt pipen-cli-gbatch, allowing --mount-as-cwd to infer workdir and outdir</li>
<li>fix: fix when running with a configuration file solely</li>
<li>fix: fix gbatch deamon workdir not created inside the pipeline working folder for gbatch cli</li>
<li>fix: fix outdir not following the pipeline name with gbatch cli</li>
<li>docs: update volume mounting syntax in docker and singularity commands</li>
<li>docs: update <code>immunopipe gbatch</code> options for running the pipeline on Google Cloud Batch Jobs</li>
<li>chore: bump pipen-cli-gbatch to 0.0.7 and pipen-report to 0.23.12</li>
<li>chore: bump biopipen to 0.34.14</li>
<li>fix(scrna_metabolic_landscape): fix report paging issue</li>
<li>docs(scrna.MarkersFinder): fix links in docs</li>
<li>fix(scrna.SeuratClusterStats): improve error handling in feature plotting when save_code (due to upgrade to ggplot2 v4)</li>
<li>feat(MarkersFinder): use <code>scplotter::MarkersPlot</code> (wrapped by <code>biopipen.utils::VizDEGs</code> to visualize marker</li>
<li>fix(scrna.CellTypeAnnotation): update logging for celltypist command execution</li>
</ul>
<h3 id="202">2.0.2<a class="headerlink" href="#202" title="Permanent link">&para;</a></h3>
<ul>
<li>fix: fix immunopipe -h/--help not working as expected</li>
<li>feat(gbatch): enhance validation and default handling in main function</li>
<li>feat(gbatch): add --mount-as-cwd option to mount a cloudpath as working directory</li>
<li>chore: bump biopipen to 0.34.10</li>
<li>docs(scrna.SeuratPreparing): enhance cell_qc parameter description in SeuratPreparing</li>
<li>docs(scrna.ModuleScoreCalculator): update link format in ModuleScoreCalculator docstring</li>
</ul>
<h3 id="201">2.0.1<a class="headerlink" href="#201" title="Permanent link">&para;</a></h3>
<ul>
<li>feat: add utility cli for gene checking and dimension verification</li>
<li>chore: make tests order work as expected</li>
<li>chore: bump biopipen to 0.34.9</li>
<li>fix(scrna.CellCellCommunication): handle numpy product attribute error</li>
<li>feat(scrna.ModuleScoreCalculator): add post mutaters functionality to allow compound modules based on added modules</li>
<li>docs(scrna.MarkersFinder): correct URL in documentation</li>
<li>feat(scrna.CellTypeAnnotation): add support for additional direct cell type annotations</li>
<li>feat(scrna.MarkersFinder): enhance enrichment plot descriptions</li>
<li>chore(scrna.CellCellCommunicationPlots): set default case to "Cell-Cell Communication"</li>
<li>feat(scrna.CellCellCommunicationPlots): add table output option for ccc data</li>
<li>chore: bump scplotter to 0.6.0 in docker image</li>
<li>chore: optimize clonal_size_data</li>
<li>feat(featurestatplot): add pos_only parameter to filter positive feature values</li>
<li>chore(enrichmentplot) remove preset label_nudge parameter in BarPlot call</li>
<li>feat: add downsample parameter to feature statistic plots</li>
<li>feat(featurestatplot) default options to TRUE to show row and column names when plot_type is heatmap-alike</li>
</ul>
<h3 id="200-cloud-support-enhanced-visualization-new-analysesfeatures-and-more">2.0.0 (Cloud support, enhanced visualization, new analyses/features, and more ...)<a class="headerlink" href="#200-cloud-support-enhanced-visualization-new-analysesfeatures-and-more" title="Permanent link">&para;</a></h3>
<h4 id="cloud-support">Cloud support<a class="headerlink" href="#cloud-support" title="Permanent link">&para;</a></h4>
<ul>
<li>feat: immunopipe can now be run on Google Cloud Batch Jobs, allowing for scalable and efficient processing of larger datasets.</li>
<li>You can either run the pipeline using the <a href="https://pwwang.github.io/pipen/scheduler/#gbatch">gbatch scheduler</a>; or</li>
<li>run the entire pipeline on Gooble Batch Jobs using <code>immunopipe gbatch</code> command.</li>
<li>See <a href="https://pwwang.github.io/immunopipe/latest/running/#run-the-pipeline-using-google-cloud-batch-jobs">Run the pipeline using Google Cloud Batch Jobs¶</a> for more details.</li>
</ul>
<h4 id="enhanced-visualization">Enhanced visualization<a class="headerlink" href="#enhanced-visualization" title="Permanent link">&para;</a></h4>
<ul>
<li>feat: <a href="https://github.com/pwwang/scplotter">scplotter</a> and <a href="https://github.com/pwwang/plotthis">plotthis</a> are now used for plotting, providing enhanced visualization capabilities and uniformity across different processes.</li>
<li>feat: default descriptions/captions are now added to plots, making them more informative.</li>
</ul>
<h4 id="enhanced-performance">Enhanced performance<a class="headerlink" href="#enhanced-performance" title="Permanent link">&para;</a></h4>
<ul>
<li>feat: the pipeline now uses <a href="https://github.com/qsbase/qs2"><code>qs2</code></a> for store the R objects, which speeds up the loading and saving of Seurat objects.</li>
<li>feat: step-wise caching (in addition to process-wise) is now supported, especially for Seurat processes, allowing for faster re-running of the pipeline by caching intermediate results and improving results reproducibility.</li>
<li>refactor: <a href="https://pwwang.github.io/immunopipe/dev/processes/ScrnaMetabolicLandscape/">MetabolicLandscapeAnalysis</a> is refactored for flexibility and performance improvement.</li>
<li>BREAKING: <a href="https://cran.r-project.org/web/packages/enrichR/vignettes/enrichR.html">enrichR</a> is retired and replaced by <a href="https://github.com/pwwang/enrichit">enrichit</a> for enrichment analysis, making it offline and more flexible. This enables the entire pipline to run without internet connection.</li>
</ul>
<h4 id="new-analysesfeatures">New analyses/features<a class="headerlink" href="#new-analysesfeatures" title="Permanent link">&para;</a></h4>
<ul>
<li>feat: the pipeline now supports <a href="https://pwwang.github.io/immunopipe/latest/processes/CellCellCommunication/">cell-cell communication analysis</a>.</li>
<li>feat: plots are supported for all cases for <a href="https://pwwang.github.io/immunopipe/latest/processes/MarkersFinder/">MarkersFinder</a> and <a href="https://pwwang.github.io/immunopipe/latest/processes/ScFGSEA/">ScFGSEA</a>, allowing plotting the markers (DEGs) and enriched pathways for all cases (e.g. all seurat clusters) in a single plot.</li>
<li>BREAKING: the <a href="https://immunarch.com">immunarch</a> package is now replaced by <a href="https://github.com/BorchLab/scRepertoire">scRepertoire</a> for more features and allowing customized clonotype definition.</li>
<li>feat: <code>envs.mutaters</code> is now supported for <a href="https://pwwang.github.io/immunopipe/latest/processes/SeuratPreparing/">SeuratPreparing</a> to allow create factor (categorical) columns in the metadata.</li>
<li>feat: <a href="https://pwwang.github.io/immunopipe/latest/processes/PseudoBulkDEG/">PseudoBulkDEG</a> is added to perform pseudo-bulk differential expression analysis.</li>
<li>feat: BCR-seq data is now supported, allowing users to analyze BCR-seq data paired with scRNA-seq data.</li>
<li>feat: Now <code>Seurat</code> object (in RDS or qs2 format) is supported as input for scRNA-seq data.</li>
<li>feat: Now loom format is supported for scRNA-seq data, allowing users to use loom files as input for the pipeline.</li>
<li>feat: add mcp server functionality to launch mcp server to help compose the configuration file.</li>
</ul>
<h4 id="house-keeping">House keeping<a class="headerlink" href="#house-keeping" title="Permanent link">&para;</a></h4>
<ul>
<li>build: docker images are now built based on the <a href="https://github.com/pwwang/biopipen"><code>biopipen</code></a> base image.</li>
<li>ci: the test workflow now caches the running intermediate files to speed up the tests.</li>
<li>docs: the citation information is now added to the documentation, allowing users to easily cite the pipeline in their publications.</li>
<li>chore(deps): biopipen is bumped to 0.34.8, which includes various bug fixes and enhancements. See the <a href="https://github.com/pwwang/biopipen/releases">biopipen releases</a> for more details.</li>
</ul>
<h3 id="144">1.4.4<a class="headerlink" href="#144" title="Permanent link">&para;</a></h3>
<ul>
<li>chore(deps): add gcc_linux-64 to Docker environment dependencies</li>
<li>docs: add input and output sections to multiple process documentation files</li>
<li>feat: add PDF output for K-means and T cell plots, enhancing report generation</li>
<li>deps: bump biopipen to 0.32.3</li>
<li>0.32.1:<ul>
<li>fix(scrna.ScFGSEA): fix case gmtfile not working</li>
<li>fix(TopExpressingGenes): add InlineNotification component to TopExpressingGenes.svelte</li>
<li>fix(scrna.SeuratClusterStats): fix kind not being added to the figure file name for plots of features</li>
<li>feat(scrna.SeuratPreparing): support percent.mt, percent.ribo, percent.hb and percent.plat for mouse</li>
</ul>
</li>
<li>0.32.2:<ul>
<li>feat: add PDF output option for SampleInfo plots</li>
<li>feat: add PDF output options for violin and scatter plots in Seurat preparation scripts</li>
<li>feat: add PDF output options for volcano, dotplot, venn, and upset plots</li>
<li>feat: add PDF output option for Enrichr plots in TopExpressingGenes script</li>
<li>feat: add PDF output options for UMAP plots in SeuratMap2Ref script; update image handling in misc.liq</li>
<li>feat: add PDF output options for cluster size distribution, shared clusters, and sample diversity plots; update plotting functions to handle multiple output formats</li>
<li>feat: add PDF output options for various Immunarch scripts; enhance reporting with downloadable PDF files</li>
<li>feat: add PDF output options for cluster size distribution, dimension plots, and feature plots; enhance reporting with downloadable PDF files</li>
<li>feat: add PDF output options for radar and bar plots; enhance reporting with downloadable PDF files</li>
<li>feat: add PDF output options for CloneResidency script; enhance reporting with downloadable PDF files</li>
<li>feat: add PDF output options for GSEA table and enrichment plots; enhance reporting with downloadable PDF files</li>
<li>feat: add PDF output options for pie charts, heatmaps, Venn plots, and UpSet plots; enhance reporting with downloadable PDF files</li>
<li>feat: add PDF output options for Enrichr plots; enhance reporting with downloadable PDF files</li>
<li>feat: add PDF output options for estimated coefficients and distribution plots; enhance reporting with downloadable PDF files</li>
</ul>
</li>
<li>0.32.3:<ul>
<li>chore: add descriptive summaries for fgsea and enrichr results</li>
</ul>
</li>
</ul>
<h3 id="143">1.4.3<a class="headerlink" href="#143" title="Permanent link">&para;</a></h3>
<ul>
<li>deps: update pipen-runinfo dependency to version 0.8.0 (pipen to 0.15.2)</li>
<li>deps: update biopipen dependency to version 0.31.4</li>
<li>fix(scrna.SeuratMap2Ref): fix refnorm not detected for NormlizeData'ed reference</li>
</ul>
<h3 id="142">1.4.2<a class="headerlink" href="#142" title="Permanent link">&para;</a></h3>
<ul>
<li>deps: add bioconductor-destiny dependency in docker environment files for ModuleScoreCalculator</li>
</ul>
<h3 id="141">1.4.1<a class="headerlink" href="#141" title="Permanent link">&para;</a></h3>
<ul>
<li>docs: update tutorial dataset information and links</li>
</ul>
<h3 id="140">1.4.0<a class="headerlink" href="#140" title="Permanent link">&para;</a></h3>
<ul>
<li>docs: update Singularity and Apptainer commands to include --writable-tmpfs flag</li>
<li>docs: allow collapsing ns/choice items in the docs for processes</li>
<li>docker: update Dockerfile to include npm cache configuration to allow the pipeline to run on read-only file system</li>
<li>tests: update SeuratPreparing config to use DoubletFinder for doublet detection</li>
<li>ci: use latest actions</li>
<li>deps: add r-clustree as a dependency in docker environment files</li>
<li>deps: update dependencies in docker environment files (python3.10, R4.3)</li>
<li>deps: bump biopipen to 0.29.0</li>
<li>fix(tcr.TCRClusterStats): fix envs.shared_clusters.heatmap_meta being broken by envs.shared_clusters.sample_order (@<a href="&#109;&#97;&#105;&#108;&#116;&#111;&#58;&#108;&#105;&#46;&#121;&#105;&#110;&#103;&#64;&#109;&#97;&#121;&#111;&#46;&#101;&#100;&#117;">&#108;&#105;&#46;&#121;&#105;&#110;&#103;&#64;&#109;&#97;&#121;&#111;&#46;&#101;&#100;&#117;</a>)</li>
<li>choir(scrna.SeuratMap2Ref): present better error message when envs.use or values of envs.MapQuery.refdata not in reference (@<a href="&#109;&#97;&#105;&#108;&#116;&#111;&#58;&#108;&#105;&#46;&#121;&#105;&#110;&#103;&#64;&#109;&#97;&#121;&#111;&#46;&#101;&#100;&#117;">&#108;&#105;&#46;&#121;&#105;&#110;&#103;&#64;&#109;&#97;&#121;&#111;&#46;&#101;&#100;&#117;</a>)</li>
<li>fix(scrna.MarkersFinder): run PrepSCTFindMarkers when needed</li>
<li>choir(scrna.SeuratClustering): use FindClusters to run for multiple resolutions</li>
<li>choir(scrna.SeuratSubClustering): use FindClusters to run for multiple resolutions</li>
<li>feat(scrna.SeuratClustering): add clustree plot (@<a href="&#109;&#97;&#105;&#108;&#116;&#111;&#58;&#108;&#105;&#46;&#121;&#105;&#110;&#103;&#64;&#109;&#97;&#121;&#111;&#46;&#101;&#100;&#117;">&#108;&#105;&#46;&#121;&#105;&#110;&#103;&#64;&#109;&#97;&#121;&#111;&#46;&#101;&#100;&#117;</a>)</li>
<li>feat(scrna.SeuratSubClustering): add clustree plot</li>
<li>tests(scrna.SeuratClusterStats): add assertion for clustree plot generation</li>
<li>deps: bump biopipen to 0.29.1</li>
<li>fix(delim.SampleInfo): fix numbers not split up when each is specified.</li>
<li>enh(delim.SampleInfo): make sizes of pie charts proportional to number of samples when each is specified</li>
<li>enh(scrna.MarkersFinder): run PrepSCTFindMarkers when necessary before calling FindMarkers</li>
<li>feat(scrna.SeuratPreparing): add option to cache Seurat object at different steps</li>
<li>feat(scrna.SeuratPreparing): allow doubletfinder to run with a different number of cores</li>
<li>chore(scrna.SeuratClustering): record PrepSCTFindMarkers command in sobj@commands</li>
<li>tests(scrna.SeuratClusterStats): use less stringent p-value cutoff for DEG/MarkersFinder</li>
<li>tests(scrna.SeuratPreparing): add doubletfinder in tests</li>
<li>deps: bump biopipen to 0.29.2</li>
<li>chore(scrna.SeuratClusterStats): use ident label length to adjust default height for feature plots</li>
<li>fix(scrna.MetaMarkers): fix seurat object not updated when expanding cases and run <code>PrepSCTFindMarkers</code> when necessary before calling meta-markers</li>
<li>fix(scrna.MarkersFinder): fix fetching command when composing the PrepSCTFindMarkers command</li>
<li>fix(scrna_metabolic_landscape): handle null values in for loop in MetabolicFeatures and MetabolicFeaturesIntraSubset for report generation</li>
<li>deps: bump biopipen to 0.30.0</li>
<li>BREAKING(scrna): move clustree plots from SeuratClustering/SeuratSubClustering to SeuratClusterStats</li>
<li>feat(scrna.CellTypeAnnotation): allow to merge/not to merge (envs.merge) the clusters with the same labels predicted</li>
<li>feat(scrna.SeuratPreparing): add scDblFinder to detect doublets</li>
<li>feat(scrna.SeuratMap2Ref): add envs.skip_if_normalized option to skip normalization if query is already normalized using the same - method as the reference</li>
<li>refactor(tcr.Immunarch): source the files for Immunarch scripts for better debugging</li>
<li>refactor(scnra.SeuratClustering): refactor the script for better debugging</li>
<li>refactor(scnra.SeuratPreparing): refactor the script for better debugging</li>
<li>fix(scrna): fix resolution expansion for SeuratClustering and SeuratSubClustering</li>
<li>fix(scrna): Fix generating PrepSCTFindMarkers command when no previous commands present</li>
<li>tests(scrna.ScFGSEA): fix unavailable urls to GMT files</li>
<li>chore(scrna.SeuratMap2Ref): optimize memory usage</li>
<li>chore(scrna.MetaMarkers): remove plugin_opts.poplog_max</li>
<li>chore(tcr.CloneResidency): improve logging when handling subjects</li>
<li>deps: bump biopipen to 0.31.3</li>
<li>enh(scrna.SeuratMap2Ref): check if reference has SCTModel if SCTransform'ed (likely prepared by old Seurat)</li>
<li>fix(tcr.CDR3AAPhyschem): use sequence from TRB chain only</li>
<li>fix(tcr.CDR3AAPhyschem): fix when chain is not available</li>
<li>fix(tcr.TCRClustering): fix for multi-chain TCRs, use TRB only if on_multi is false</li>
<li>fix(tcr.TCRClustering): fix when chain is not available</li>
</ul>
<h3 id="139">1.3.9<a class="headerlink" href="#139" title="Permanent link">&para;</a></h3>
<ul>
<li>docs: update docs for TCellSelection to avoid confusion</li>
<li>deps: bump biopipen to 0.27.9</li>
<li>feat(tcr.TCRClusterStats): add sample_order to set sample order on heatmap and cluster_rows to switch row clustering on/off</li>
</ul>
<h3 id="138">1.3.8<a class="headerlink" href="#138" title="Permanent link">&para;</a></h3>
<ul>
<li>docs: remove -w option for apptainer/singularity as no writing is necessary since pipen-board 0.15.1</li>
<li>deps: update biopipen to version 0.27.8</li>
<li>fix(scrna.SeuratClusterStats): fix selected columns not unique for stats</li>
<li>feat(scrna.SeuratMap2Ref): allow non-SCTransform'ed reference</li>
<li>feat(scrna.SeuratMap2Ref): allow splitting query object for mapping (pwwang/immunopipe#61)</li>
<li>deps: update pipen-board to version 0.15.1 (allow configuration file path in the URL box on Web UI)</li>
</ul>
<h3 id="137">1.3.7<a class="headerlink" href="#137" title="Permanent link">&para;</a></h3>
<ul>
<li>ci: fix docker images building when no essential changes made</li>
</ul>
<h3 id="136">1.3.6<a class="headerlink" href="#136" title="Permanent link">&para;</a></h3>
<ul>
<li>ci: fix deploy workflow (#59)</li>
<li>ci: add README.md to tests-output branch</li>
<li>ci: fix test/test workflow</li>
<li>tests: add <code>make test</code></li>
<li>tests: init test data preparation</li>
<li>tests: add test for ImmunarchLoading</li>
<li>tests: add tests for SeuratPreparing</li>
<li>tests: Update configs for SeuratPreparing test to subset cells so tests can run on CI</li>
<li>tests: update SeuratPreparing test to disable export</li>
<li>tests: add tests for SeuratClusteringOfAllCells/SeuratClustering</li>
<li>docs: update installation instructions (@<a href="&#109;&#97;&#105;&#108;&#116;&#111;&#58;&#115;&#116;&#101;&#105;&#110;&#46;&#109;&#97;&#114;&#105;&#97;&#109;&#64;&#109;&#97;&#121;&#111;&#46;&#101;&#100;&#117;">&#115;&#116;&#101;&#105;&#110;&#46;&#109;&#97;&#114;&#105;&#97;&#109;&#64;&#109;&#97;&#121;&#111;&#46;&#101;&#100;&#117;</a>)</li>
<li>deps: bump biopipen to version 0.27.7 (0.27.5-0.27.7)</li>
<li>fix(scrna.SeuratClusterStats): fix color palette for ridge plots (@<a href="&#109;&#97;&#105;&#108;&#116;&#111;&#58;&#115;&#116;&#101;&#105;&#110;&#46;&#109;&#97;&#114;&#105;&#97;&#109;&#64;&#109;&#97;&#121;&#111;&#46;&#101;&#100;&#117;">&#115;&#116;&#101;&#105;&#110;&#46;&#109;&#97;&#114;&#105;&#97;&#109;&#64;&#109;&#97;&#121;&#111;&#46;&#101;&#100;&#117;</a>)</li>
<li>feat(scrna.SeuratPreparing): add envs.cell_qc_per_sample to filter cells before merging instead after</li>
<li>fix(scrna_metabolic_landscape.MetabolicFeatures): fix return value of groups with less than 5 cells in <code>do_one_group</code></li>
<li>fix(scrna_metabolic_landscape): fix mutaters not working.</li>
<li>fix(scrna_metabolic_landscape.MetabolicFeatures/MetabolicFeaturesIntraSubset): skip groups with less than 5 cells in <code>do_one_group</code> and save a warning file under the case</li>
<li>chore: fix typo in class name <code>ExprImpution</code> to <code>ExprImputation</code></li>
</ul>
<h3 id="135">1.3.5<a class="headerlink" href="#135" title="Permanent link">&para;</a></h3>
<ul>
<li>ci/test: add tests in CI and deploy output in a different branch</li>
<li>deps: bump biopipen to 0.27.4</li>
<li>choir(delim.SampleInfo): add alpha to the colors of the plots using biopipen color pallete</li>
<li>docs(tcr/scrna/scrna_metabolic_landscape): update links of images in docs</li>
</ul>
<h3 id="134-post">1.3.4-post<a class="headerlink" href="#134-post" title="Permanent link">&para;</a></h3>
<ul>
<li>ci/test: init ci for tests</li>
<li>docs: introduce versioning for docs</li>
</ul>
<h3 id="134">1.3.4<a class="headerlink" href="#134" title="Permanent link">&para;</a></h3>
<ul>
<li>deps: bump biopipen to 0.27.3</li>
<li>deps: bump pipen-poplog to 0.1.2 (quick fix for populating logs when job fails)</li>
<li>deps: temporary fix copier breaks with pyyaml-include v2 (copier-org/copier#1568)</li>
<li>choir(scrna.ScFGSEA): Skip cases when no cells found (#50)</li>
<li>choir(scrna.MarkersFinder): Skip cases when no cells found (#50)</li>
<li>choir(scrna.MetaMarkers): Skip cases when no cells found (#50)</li>
<li>feat(scrna.SeuratPreparing): support <code>DoubletFinder</code> (#52)</li>
</ul>
<h3 id="133">1.3.3<a class="headerlink" href="#133" title="Permanent link">&para;</a></h3>
<ul>
<li>deps: temporary fix copier breaks with pyyaml-include v2 (copier-org/copier#1568)</li>
<li>docs: update FAQ.md with instructions for running pipeline on a cluster</li>
<li>deps: bump biopipen to 0.27.2</li>
<li>fix(scrna.RadarPlots): fix mutaters not working</li>
<li>feat(tcr.CloneResidency): support <code>envs.upset_ymax</code> to set the max value of y axis in upset bar plot.</li>
</ul>
<h3 id="132">1.3.2<a class="headerlink" href="#132" title="Permanent link">&para;</a></h3>
<ul>
<li>deps: bump pipen to 0.14.5</li>
<li>deps: add r-complexupset package to environment.yml and environment_full.yml for CloneResidency</li>
<li>deps: pin tensorflow to 2.15 for TESSA</li>
<li>deps: bump biopipen to 0.27.1</li>
<li>depr(scrna.MarkersFinder): remove <code>envs.use_presto</code> as it's used by Seurat v5 by default</li>
<li>enh(tcr.CloneResidency): support log scale for y axis of upset bar plots</li>
<li>enh(scrna.SeuratClusterStats): allow to rotate labels in circos plot (pwwang/immunopipe#48) @<a href="&#109;&#97;&#105;&#108;&#116;&#111;&#58;&#108;&#105;&#46;&#121;&#105;&#110;&#103;&#64;&#109;&#97;&#121;&#111;&#46;&#101;&#100;&#117;">&#108;&#105;&#46;&#121;&#105;&#110;&#103;&#64;&#109;&#97;&#121;&#111;&#46;&#101;&#100;&#117;</a></li>
<li>enh(scrna.SeuratClusterStats): use <code>pal_biopipen</code> for ident colors in circos plot</li>
<li>fix(scrna.CellsDistribution): fix the row order of the heatmaps</li>
<li>fix(scrna.SeuratClusterStats): fix when <code>envs.split-by</code> is specified</li>
<li>feat(scrna.CellsDistribution): support <code>envs.prefix_each</code></li>
<li>feat(scrna.MarkersFinder): allow to set max number of genes to plot in dotplots</li>
<li>feat(scrna.MarkersFinder): support setting detailed arguments for overlapping plots</li>
<li>feat(scrna.MarkersFinder): support <code>envs.prefix_group</code></li>
<li>feat(scrna.ScFGSEA): support <code>envs.prefix_each</code></li>
<li>feat(scrna.RadarPlots): support <code>envs.prefix_each</code> and <code>envs.subset</code></li>
<li>choir(scrna.SeuratClusterStats): use logger instead of print for log messages</li>
<li>choir(tcr.TCRClustering): print session info for <code>clustcr</code> script</li>
<li>choir(scrna.MarkersFinder): flatten toc when no <code>section</code> and no <code>ident-1</code> specified</li>
<li>docs: add more detailed docs for <code>envs.section</code> for multiple processes</li>
<li>BREAKING(scrna.SeuratMap2Ref): rename envs.name to envs.ident so envs.MapQuery.refdata is not - required anymore. It will be inferred from envs.ident and envs.use. @<a href="&#109;&#97;&#105;&#108;&#116;&#111;&#58;&#108;&#105;&#46;&#121;&#105;&#110;&#103;&#64;&#109;&#97;&#121;&#111;&#46;&#101;&#100;&#117;">&#108;&#105;&#46;&#121;&#105;&#110;&#103;&#64;&#109;&#97;&#121;&#111;&#46;&#101;&#100;&#117;</a></li>
</ul>
<h3 id="131">1.3.1<a class="headerlink" href="#131" title="Permanent link">&para;</a></h3>
<ul>
<li>deps: bump pipen to 0.14.3</li>
<li>deps: pin ggplot2 to 3.4 for docker due to breaking changes of 3.5</li>
<li>deps: bump biopipen to 0.26.2</li>
<li>deps: bump datar-pandas to 0.5.5 to dismiss deprecated warnings</li>
<li>fix(utils.misc.R): replace latin and greek characters with closest ascii chars for slugify()</li>
<li>feat(scrna.TopExpressingGenes): support <code>subset</code></li>
<li>fix(scrna.CellsDistribution): fix the row order of the heatmaps.</li>
<li>enh(tcr.CloneResidency): add legend for multiplets in upset plots.</li>
<li>feat(scrna.SeuratClusterStats): add circos plot for cell composition stats (#46).</li>
</ul>
<h3 id="130">1.3.0<a class="headerlink" href="#130" title="Permanent link">&para;</a></h3>
<ul>
<li>deps: bump pipen to 0.14.1</li>
<li>deps: bump pipen-report to 0.18.2</li>
<li>deps: bump biopipen to 0.26.0</li>
<li>fix(scrna.CellTypeAnnotation): keep factor meta data when input and output are RDS for celltypist</li>
<li>deps: bump datar to 0.15.4 (support pandas 2.2)</li>
<li>fix(utils.single_cell.R): fix <code>immdata_from_expanded</code> missing other data columns</li>
<li>fix(tcr.Immunarch): fix <code>mutaters</code> not working when no subset is set</li>
<li>fix(scrna.CellsDistribution): fix <code>hm_devpars</code> not working</li>
<li>fix(scrna.CellsDistribution): fix multiple <code>cells_by</code> columns and speed up plotting</li>
<li>choir(tcr.CloneResidency): mark singletons in Venn diagrams more clear</li>
<li>fix(scrna.RadarPlots): fix the order of groups on radar plots</li>
<li>choir(scrna.RadarPlots): transpose the count/percentage table to save to files</li>
<li>fix(scrna.MarkersFinder): fix generating report json file when no significant genes found</li>
<li>choir(scrna.MarkersFinder): Plot maximum 20 genes in dotplots</li>
<li>choir(scrna.MarkersFinder): Do not convert dashes in case names to dots</li>
<li>see more at <a href="https://github.com/pwwang/biopipen/releases/tag/0.26.0">https://github.com/pwwang/biopipen/releases/tag/0.26.0</a></li>
</ul>
<h3 id="120">1.2.0<a class="headerlink" href="#120" title="Permanent link">&para;</a></h3>
<ul>
<li>docs: update FAQs to align with Seurat v5</li>
<li>docs: add image from manuscript to README.md</li>
<li>docs: center the flowchart image in README.md</li>
<li>docs: mention <code>celltypist</code> model prep in preparing input data</li>
<li>deps: bump pipen to 0.13.2</li>
<li>deps: bump biopipen to 0.25.2:</li>
<li>scrna.MarkersFinder: allow to cache <code>FindAllMarkers</code> results</li>
<li>scrna.CellTypeAnnotation: support <code>celltypist</code> (pwwang/biopipen#111)</li>
<li>scrna.SeuratSubClustering: add <code>envs_depth = 1</code> to replace whole <code>envs.cases</code> when new case assigned</li>
<li>scrna_metabolic_landscape.MetabolicPathwayHeterogeneit): fix output directory path is not slugified</li>
<li>tcr.Immunarch: change case filling log to debug level</li>
</ul>
<h3 id="111">1.1.1<a class="headerlink" href="#111" title="Permanent link">&para;</a></h3>
<ul>
<li>deps: Bump biopipen to 0.24.2</li>
<li>chore: use internal slugify instead of slugify library</li>
<li>tcr.Immunarch: fix spectratyping output file extension is not png</li>
<li>scrna.SeuratPreparing: fix displaying filters in report</li>
<li>scrna.SeuratPreparing: fix logging Seurat procedure arguments</li>
</ul>
<h3 id="110">1.1.0<a class="headerlink" href="#110" title="Permanent link">&para;</a></h3>
<ul>
<li>docs: update table in gallery</li>
<li>deps: use <a href="https://github.com/pwwang/pipen-poplog">pipen-poplog</a> to populate job logs to pipeline running log</li>
<li>deps: bump biopipen to 0.24. Hights:</li>
<li>scrna.ScFGSEA: add subset to filter cells (pwwang/biopipen#112) @yuey11</li>
<li>scrna.SeuratClustering/SeuratSubClustering: cache Seurat procedures step by step (#40) @xyfqwlzoe</li>
<li>tcr.Immunarch: add <code>plot_type</code> to support boxplots for diversity metrics</li>
<li>see more at <a href="https://github.com/pwwang/biopipen/releases/tag/0.24.0">https://github.com/pwwang/biopipen/releases/tag/0.24.0</a></li>
</ul>
<h3 id="105">1.0.5<a class="headerlink" href="#105" title="Permanent link">&para;</a></h3>
<ul>
<li>change: do not rescale gene expression in <code>TCellSelection</code> any more</li>
<li>fix: fix column names of indicators not aligned with <code>indicator_genes</code></li>
<li>feat: add feature plots in <code>TCellSelection</code></li>
<li>deps: bump biopipen to 0.23.8</li>
<li>scrna.SeuratPreparing: log Seurat procedure arguments</li>
<li>scrna.ScFGSEA: add <code>subset</code> to filter cells (pwwang/biopipen#112)</li>
</ul>
<h3 id="104">1.0.4<a class="headerlink" href="#104" title="Permanent link">&para;</a></h3>
<ul>
<li>deps: bump biopipen to 0.23.7</li>
<li>scrna.SeuratPreparing: update log message for transformation/scaling step</li>
<li>scrna_metabolic_landscape.MetabolicPathwayHeterogeneity: add utils.gsea script source to support <code>localizeGmtfile</code></li>
</ul>
<h3 id="103">1.0.3<a class="headerlink" href="#103" title="Permanent link">&para;</a></h3>
<ul>
<li>deps: add r-seuratdisk dependency to conda env files. @yuey11</li>
<li>deps: pin r-matrixstats to 1.1.0 in conda env files to fix <code>useNames = NA</code> error. @yuey11</li>
<li>refactor: optimize configuration file validation</li>
<li>deps: bump biopipen to 0.23.6</li>
<li>feat: support url for gmtfile wherever GSEA is performed (pwwang/biopipen#113)</li>
<li>tcr.Immunarch: add error message for empty filtered/subset data in diversity</li>
<li>scrna.SeuratPreparing: correct description of default assay in docstr</li>
<li>scrna.SeuratPreparing: run also the normal normalization procedures when <code>SCTransform</code> is used (useful for visualization purposes on RNA assay)</li>
<li>scrna.ModuleScoreCalculator: document the names added by cell cycle score (#34)</li>
<li>scrna.SeuratPreparing: support sample names as <code>reference</code> for <code>IntegrateLayers</code></li>
</ul>
<h3 id="102">1.0.2<a class="headerlink" href="#102" title="Permanent link">&para;</a></h3>
<ul>
<li>deps: add bioconductor-glmgampoi to conda env files (#33)</li>
<li>docs: correct the Seurat object assay description</li>
<li>deps: bump biopipen to 0.23.5</li>
<li>fix: fix when no enriched items found for <code>scrna.MarkersFinder</code>, <code>scrna.MetaMarkers</code> and <code>scrna.TopExpressingGenes</code></li>
<li>scrna.SeuratClusterStats: fix when <code>frac</code> or <code>frac_ofall</code> is true and no <code>group-by</code> nor <code>split-by</code> is specified for <code>stats</code></li>
<li>utils.gsea.R: fix when no enriched items found for <code>runEnrichr</code></li>
<li>scrna_metabolic_landscript: fix adding report when <code>ncores</code> &gt; 1</li>
</ul>
<h3 id="101">1.0.1<a class="headerlink" href="#101" title="Permanent link">&para;</a></h3>
<ul>
<li>docs: add gallery section to README.md</li>
<li>change: set default <code>nstart</code> of kmeans to 25 in <code>TCellSelection</code></li>
<li>deps: add <code>r-hdf5r</code> in conda env files to support <code>Read_10x_h5</code> from Seurat. @yuey11</li>
<li>deps: bump biopipen to 0.23.4</li>
<li>scrna.TopExpressingGenes: fix colnames while pulling average expression</li>
<li>scrna.CellsDistribution: fix when <code>cells_by</code> has multiple column names</li>
<li>scrna.CellTypeAnnotation: fix the order of the clusters for <code>direct</code> method</li>
<li>scrna.SeuratClusterStats: add <code>position</code> options for bar plots for stats</li>
<li>scrna.RadarPlots: add <code>colors</code> to set the colors of the loops in radar and bar plots</li>
<li>tcr.Immunarch: add <code>split_by</code> and <code>split_order</code> to put subplots together in one single plots</li>
</ul>
<h3 id="100">1.0.0<a class="headerlink" href="#100" title="Permanent link">&para;</a></h3>
<h4 id="highlights">Highlights<a class="headerlink" href="#highlights" title="Permanent link">&para;</a></h4>
<ul>
<li>feat: support <code>Seurat</code> v5 (integration is now down by <code>Seurat::IntegrateLayers</code>)</li>
<li>feat: support supervised clustering (mapping cells to reference by <code>Seurat</code>)</li>
<li>feat: support dataset with scRNA-seq data only (no scTCR-seq data)</li>
<li>feat: support diffusion map calculation (by <code>ModuleScoreCalculator</code>)</li>
<li>feat: support subclassing to cluster subsets of cells (by <code>SeuratSubClustering</code>)</li>
<li>feat: allow to ignore TCR data in <code>TCellSelection</code> and pass kmeans arguments</li>
<li>feat: allow to set multiple resolutions (<code>envs.FindClusters.resolution</code>) in <code>SeuratClustering</code>/<code>SeuratClusteringOfTCells</code></li>
<li>change: change unsuperved cluster labels to <code>c1</code>, <code>c2</code>, ... in <code>SeuratClustering</code> by default</li>
<li>docs: add gallery, which contains real-world examples of datasets from publications</li>
</ul>
<h4 id="breaking-changes">Breaking changes<a class="headerlink" href="#breaking-changes" title="Permanent link">&para;</a></h4>
<ul>
<li>change: rename <code>SeuratMetadataMutater</code> to <code>IntegratingTCR</code></li>
<li>change: rename <code>SeuratClusteringOfTCells</code> to <code>SeuratClustering</code></li>
<li>change: rename <code>TCRClusters2Seurat</code> to <code>IntegratingTCRClusters</code></li>
<li>refactor: make <code>SeuratClustering</code> (instead of <code>SeuratClusteringOfAllCells</code>) work for all cells when all are T cells</li>
<li>change: move data preparation and integration from <code>SeuratClustering</code> to <code>SeuratPreparing</code></li>
<li>change: default <code>mode</code> of <code>ImmunarchLoading</code> to <code>paired</code> (instead of <code>single</code>), which requires both alpha and beta chains (instead of beta chain only) to define a clonotype</li>
<li>change: default <code>dbs</code> for enrichment analysis wherever applies to <code>KEGG_2021_Human</code> and <code>MSigDB_Hallmark_2020</code></li>
</ul>
<h4 id="changes">Changes<a class="headerlink" href="#changes" title="Permanent link">&para;</a></h4>
<ul>
<li>feat: make <code>TopExpressingGenes</code> optional</li>
<li>feat: add <code>validate_config</code> to validate configuration schematically</li>
</ul>
<h4 id="features">Features<a class="headerlink" href="#features" title="Permanent link">&para;</a></h4>
<ul>
<li>feat(SeuratPreparing): allow to filter genes directly (by specifying <code>envs.gene_qc.excludes</code>)</li>
<li>feat(SeuratClusterStats): add <code>ngenes</code> to plot the number of genes expressed in each cluster</li>
<li>feat(SeuratClusterStats): add <code>barplot</code> for features and allow aggregation of features</li>
<li>feat(SeuratClusterStats): add <code>envs.mutaters</code> to mutate meta data</li>
<li>feat(SeuratClusterStats): add histograms to plot number of cells against another variable</li>
<li>feat(SeuratClusterStats): Add <code>frac_ofall</code> and <code>transpose</code> for stats to calculate fraction within group or against all cells, and transpose ident and group, respectively</li>
</ul>
<h4 id="dependencies">Dependencies<a class="headerlink" href="#dependencies" title="Permanent link">&para;</a></h4>
<ul>
<li>deps: add <code>r-presto</code> to conda environment files to support using presto to fastly find markers</li>
<li>deps: add <code>bioconductor-destiny</code> to conda environment file to support add diffusion map components in ModuleScoreCalculator</li>
<li>deps: add <code>r-harmony</code> to support harmony integration by Seurat v5 in conda env file</li>
<li>deps: add <code>r-sf</code> to conda env file</li>
<li>deps: remove <code>vdjtools</code> from conda env files</li>
<li>deps: bump <code>pipen-report</code> to <a href="https://github.com/pwwang/pipen-report/releases/tag/0.16.3">0.16.3</a></li>
<li>deps: bump <code>biopipen</code> to <a href="https://github.com/pwwang/biopipen/releases">0.23.3</a>. Hightlight changes:</li>
<li>scrna.MarkersFinder: Add <code>envs.use_presto</code> to use presto to speed up finding markers</li>
<li>scrna.SeuratPreparing: Set envs.gene_qc.min_cells to 0 by default (instead of 3)</li>
<li>scrna.ScFGSEA: Allow to ignore small group when fgsea fails due to all NAs for pre-ranks</li>
<li>scrna.CellsDistribution: Allow to order clusters by envs.cluster_orderby</li>
<li>scrna.CellsDistribution: Add heatmaps</li>
<li>tcr.CloneResidency: Make section works in report</li>
<li>tcr.Immunarch: Support paired chain data for VJ conjuction plots</li>
<li>tcr.TESSA: Change envs.assay to None to use default assay of Seurat object</li>
<li>scrna.SeuratClusterStats: Add <code>avgheatmap</code> to plot more elegant heatmap for average gene expressions</li>
<li>scrna.SeuratClusterStats: Fix ident not working for dimplots</li>
<li>scrna.SeuratClusterStats: Add <code>cluster_orderby</code> to order clusters for features</li>
<li>scrna.SeuratClusterStats: Add <code>na_group</code> to keep NA values in <code>group-by</code></li>
<li>utils.mutate_helpers: Change arguments <code>id_col</code> and <code>compare_col</code> of <code>paired</code> to <code>id</code> and <code>compare</code>, respectively</li>
<li>utils.mutate_helpers: Fix that subset can't be an expression for expanded family</li>
<li>utils.mutate_helpers: Add <code>top</code> to select top entities (e.g clones)</li>
<li>scrna.RadarPlots: Add breakdown and test to break down the cell distribution and run statistic test on the fractions</li>
</ul>
<h3 id="0112">0.11.2<a class="headerlink" href="#0112" title="Permanent link">&para;</a></h3>
<ul>
<li>docs: move <code>Immunarch</code> to the later position in process list</li>
<li>docs: Use <code>master</code> tag in getting-started</li>
</ul>
<h3 id="0111">0.11.1<a class="headerlink" href="#0111" title="Permanent link">&para;</a></h3>
<ul>
<li>chore: change line length to 88 for flake8</li>
<li>chore: dismissing warning about wasting columns for <code>SeuratClusteringOfTCells</code></li>
<li>docs: update CHANGELOG.md with missing changes of last version</li>
<li>docs: add version of renaming <code>envs.tcell_indicator</code> to <code>envs.tcell_selector</code></li>
<li>docs: remove unused doc files</li>
<li>docs: add metadata illustration</li>
<li>deps: bump biopipen to 0.22.8. Highlights:</li>
<li>deps: bump pipen-board to 0.13.10 (pipen-report to 0.16.2)</li>
<li>CellsDistribution: Don't add rownames to the output table file</li>
<li>MarkersFinder (ClusterMarkers/ClusterMarkersOfAllCells): Optimize to use <code>FindAllMarkers</code> if <code>ident.1</code> is not specified</li>
<li>SeuratClusterStats: Fix path of expression table file</li>
<li>CellTypeAnnotation: Allow using <code>NA</code> to exclude clusters from output <code>Seurat</code> object</li>
<li>utils.mutate_helpers: Return ids only when subset is true and group is not <code>NA</code> for <code>uniq = TRUE</code> in <code>expanded</code>, <code>collapsed</code>, <code>emerged</code> and <code>vanished</code></li>
</ul>
<h3 id="0110">0.11.0<a class="headerlink" href="#0110" title="Permanent link">&para;</a></h3>
<ul>
<li>deps: update biopipen to 0.22.1, highlights:</li>
<li>add V-J junction circos plots to <code>Immunarch</code> process</li>
<li>add cache option to cache the clustering results if nothing changed except ncores, to <code>SeuratClustering</code> process</li>
<li>add dot plots to <code>MarkersFinder</code> (<code>ClusterMarkersOfAllCells</code>, <code>ClusterMarkers</code>) process</li>
<li>save exported table with only necessary columns for <code>CellsDistribution</code> process</li>
<li>add <code>descr</code> to describe cases cases in report for <code>CellsDistribution</code> process</li>
<li>add <code>subset</code> for dimplots in <code>SeuratClusterStats</code> process</li>
<li>use a new palette (<code>biopipen</code>) for related processes</li>
<li>optimize report rendering (using <code>render_job()</code> filter from <code>pipen-report</code>)</li>
<li>change metacols to extracols so essential columns get exported for <code>ImmunarchLoading</code> process</li>
<li>add cache option to cache the clustering results if nothing changed except ncores for <code>SeuratClustering</code> (<code>SeuratClusteringOfAllCells</code>) process</li>
<li>see more at <a href="https://github.com/pwwang/biopipen/releases/tag/0.22.0">https://github.com/pwwang/biopipen/releases/tag/0.22.0</a> and <a href="https://github.com/pwwang/biopipen/releases/tag/0.22.1">https://github.com/pwwang/biopipen/releases/tag/0.22.1</a></li>
<li>deps: update pipen-report to 0.16, highlights:</li>
<li>scroll anchor into view on the page</li>
<li>build report page when each process is done, instead of the whole pipeline</li>
<li>see more at <a href="https://github.com/pwwang/pipen-report/releases/tag/0.16.0">https://github.com/pwwang/pipen-report/releases/tag/0.16.0</a></li>
<li>change: remove <code>Immunarch2VDJtools</code> and <code>VJUsage</code> processes (vj usage analysis can be done in <code>Immunarch</code> process)</li>
<li>change: change <code>tcell_indicator</code> to <code>tcell_selector</code> in <code>TCellSelection</code> process</li>
<li>enhance: provide better error message when none barcode matches from RNA and TCR data for <code>TCRClustering</code> process</li>
<li>docs: add memory usage reduction tips in FAQ</li>
<li>chore: dismiss warnings of wasted input columns for multiple processes</li>
</ul>
<h3 id="0101">0.10.1<a class="headerlink" href="#0101" title="Permanent link">&para;</a></h3>
<ul>
<li>chore: update pipeline description to include version in the logs</li>
<li>fix: add fc-cache command to Dockerfile to solve <code>Fontconfig error</code></li>
<li>docker: optimize building full image based off the base image</li>
</ul>
<h3 id="0100">0.10.0<a class="headerlink" href="#0100" title="Permanent link">&para;</a></h3>
<ul>
<li>docker: lock r-matrix version to 1.6_1 for compatibility</li>
<li>docs: adopt mkdocs-rtd 0.0.10 (add scrollbar to the table of contents)</li>
<li>deps: bump biopipen to 0.21.1</li>
<li>use <code>r-logger</code> for logging in R scripts</li>
<li>docs: fix internal references in API docs</li>
<li>deps: bump pipen-board to 0.13.6</li>
<li>SampleInfo: refactor data subset logic using <code>subset</code> instead of <code>distinct</code></li>
<li>Immunarch: add <code>in.metafile</code> to allow other meta info (i.e. seurat clusters) for future subsetting (#22)</li>
<li>Immunarch: fix empty groups in diversity plot after subsetting</li>
<li>Immunarch: allow <code>subset</code> to subset cells for analyses</li>
<li>Immunarch: allow <code>separate_by</code> also works on other diversity plots</li>
<li>Immunarch: add <code>ymin</code> and <code>ymax</code> to align diversity plots by <code>separate_by</code></li>
<li>Immunarch: add <code>ncol</code> to specify # columns in the combined plots</li>
<li>RadarPlots: fix <code>envs.order</code> not working</li>
<li>MarkersFinder: add <code>overlap</code> to find overlapping markers between cases (#24)</li>
<li>MarkersFinder: allow <code>subset</code> to subset cells for analyses</li>
<li>MarkersFinder: add dot plots for significant markers</li>
<li>CellsDistribution: allow multiple columns for <code>cells_by</code></li>
<li>CellsDistribution: allow <code>subset</code> to subset cells for analyses</li>
<li>utils.mutate_helpers.R: add <code>include_emerged</code> for <code>expanded()</code> and <code>include_vanished</code> for <code>collapsed()</code></li>
</ul>
<h3 id="093">0.9.3<a class="headerlink" href="#093" title="Permanent link">&para;</a></h3>
<ul>
<li>deps: Bump biopipen to 0.20.7</li>
<li>deps: Bump pipen-board to 0.13.4</li>
<li>ClusterMarkers/ClusterMarkersOfAllCells: Choose avg_log2FC &gt; 0 markers by default</li>
<li>MarkersFinder: Allow to set assay and set assay to <code>RNA</code> by default</li>
<li>CellsDistribution: Add venn/upset plot for overlapping cell groups in different cases</li>
<li>SampleInfo: Add <code>distinct</code> to case to perform stats on distinct records</li>
</ul>
<h3 id="092">0.9.2<a class="headerlink" href="#092" title="Permanent link">&para;</a></h3>
<ul>
<li>➕ Add <code>r-ggnewscale</code> as dependency for <code>CDR3AAPhyschem</code> in docker image</li>
<li>⬆️ Bump biopipen to 0.20.5</li>
<li>🧱 CloneResidency: Integrate RNA data to allow more flexible analysis (i.e. within specific seurat clusters)</li>
<li>🏗️ CloneResidency: Rename envs.sample_groups to envs.section to be consistent with other processes</li>
<li>📝 ScFGSEA: Remove the link in the summary of the docstring (since they are not transformed in the report)</li>
<li>🎨 CDR3AAPhyschem: Give better error message when wrong group items are given</li>
<li>⬆️ Bump pipen-board to 0.13.3</li>
<li>Add items automatically when blurred for list options</li>
<li>Add other sections to description on the UI for processes</li>
</ul>
<h3 id="091">0.9.1<a class="headerlink" href="#091" title="Permanent link">&para;</a></h3>
<ul>
<li>🐛 Fix docstring for <code>RadarPlots</code></li>
<li>➕ Add <code>pipen-diagram</code> as dependency</li>
<li>➕ Set <code>pipen-runinfo</code> as optional</li>
<li>⬆️ Bump biopipen to 0.20.4</li>
<li>📝 Update version in docs</li>
</ul>
<h3 id="090">0.9.0<a class="headerlink" href="#090" title="Permanent link">&para;</a></h3>
<h4 id="housekeeping-and-docs">Housekeeping and docs<a class="headerlink" href="#housekeeping-and-docs" title="Permanent link">&para;</a></h4>
<ul>
<li>Bump biopipen to 0.20.3 (pipen to 0.12)</li>
<li>Use <a href="https://github.com/pwwang/pipen-cli-ref"><code>pipen-cli-ref</code></a> to generate API for processes (it uses docstring of the process class so that we don't need to maintain two copies of docs)</li>
</ul>
<h4 id="fixedenhanced">Fixed/Enhanced<a class="headerlink" href="#fixedenhanced" title="Permanent link">&para;</a></h4>
<ul>
<li>Make <code>/data</code> directory in container, so it can be mounted</li>
<li>Fix a bug when a single gene provided to <code>indicator_genes</code> in <code>TCellSelection</code></li>
<li>Move <code>ModuleScoreCalculator</code> before clustering so that the scores can be used in <code>vars.to.regress</code> of <code>SCTransform</code> while clustering</li>
<li>Set default assay to RNA in case module scores only caculated using integrated features in <code>ModuleScoreCalculator</code></li>
<li>Improve QC plots in <code>SeuratPreparing</code> by marking the cells that are removed in the plots instead of doing before/after plots</li>
<li>Fix type annotation for envs.features_defaults.ncol in docstring for <code>SeuratPreparing</code> (causing <code>pipen-board</code> not converting to int)</li>
<li>Fix the cluster order in pie charts for <code>CellsDistribution</code></li>
<li>Fix the cluster order in pie charts for <code>SeuratClusterStats</code></li>
<li>Fix order in pie charts for <code>SampleInfo</code></li>
<li>Fix docstring for <code>envs.div.args</code> of <code>Immunarch</code> (more clear description of method)</li>
<li>Allow mutiple columns in the file for <code>envs.features_defaults.features</code> in <code>SeuratClusterStats</code></li>
<li>Allow order to be optional for <code>CloneResidency</code> (errored when not provided)</li>
<li>Add number of clusters at the end of log for <code>SeuratClusteringOfAllCells</code>/<code>SeuratClusteringOfTCells</code></li>
<li>Add stricter checker for input file (#13)</li>
<li>Indicate the case name in logs when pie is enabled for <code>group-by</code> in <code>SeuratClusterStats</code></li>
<li>Allow to skip overlap and gene usage analyses by setting method to <code>none</code> for <code>Immunoarch</code> (#11, #12)</li>
<li>Don't cluster on heatmap when there are only 2 samples for <code>TCRClusterStats</code> (#11)</li>
<li>Import Seurat explictly to avoid satijalab/seurat#2853 in <code>MetabolicFeatures</code></li>
<li>Fix when NA values in data for heatmap in <code>MetabolicPathwayActivity</code></li>
<li>Fix error when no significant pathways selected in <code>MetabolicPathwayHeterogeneity</code></li>
<li>Give better error message in CellsDistribution if group value not found for <code>CellsDistribution</code> (#16)</li>
<li>Try including more genes (even though insignificant) in volcano plot for <code>MarkersFinder</code>/<code>ClusterMarkers</code>/<code>ClusterMarkersOfAllCells</code> (#17)</li>
<li>Add margins to volcano plot for <code>MarkersFinder</code>/<code>ClusterMarkers</code>/<code>ClusterMarkersOfAllCells</code></li>
<li>Fix when <code>envs.cell_qc</code> is <code>None</code> (not provided) for <code>SeuratPreparing</code></li>
<li>Fix <code>ident</code> in cases of <code>envs.dimplots</code> not working for <code>SeuratClusterStats</code></li>
</ul>
<h4 id="added">Added<a class="headerlink" href="#added" title="Permanent link">&para;</a></h4>
<ul>
<li>Add <code>ClusterMarkersOfAllCells</code> and <code>TopExpressingGenesOfAllCells</code> and set them as optional</li>
<li>Add dim plots in <code>SeuratClusterStats</code> to overlay TCR presence/absence of cells (#14)</li>
</ul>
<h4 id="breaking-changes-090">Breaking changes-0.9.0<a class="headerlink" href="#breaking-changes-090" title="Permanent link">&para;</a></h4>
<ul>
<li>Rename <code>TCRClusteringStats</code> to <code>TCRClusterStats</code> (#15)</li>
</ul>
<h3 id="083">0.8.3<a class="headerlink" href="#083" title="Permanent link">&para;</a></h3>
<ul>
<li>📝 Fix typos in docs</li>
<li>📝 Add links to some optional input files (#9, 5)</li>
<li>🔨 Add apptainer to docker entry.sh (#9, 6)</li>
<li>💄 Adjust process order in reports (#9, 1)</li>
<li>⬆️ Bump pipen-report to 0.13.1 (#9, 2)</li>
</ul>
<h3 id="082">0.8.2<a class="headerlink" href="#082" title="Permanent link">&para;</a></h3>
<ul>
<li>Bump biopipen to 0.18.3 to fix when either ident is empty for <code>MarkersFinder</code></li>
</ul>
<h3 id="081">0.8.1<a class="headerlink" href="#081" title="Permanent link">&para;</a></h3>
<ul>
<li>Bump biopipen to 0.18.2 to fix a bug when the min length of CDR3 seqs &gt; 12 for <code>CDR3AAphyschem</code></li>
</ul>
<h3 id="080">0.8.0<a class="headerlink" href="#080" title="Permanent link">&para;</a></h3>
<h4 id="housekeeping-and-docs-updates">Housekeeping and docs updates<a class="headerlink" href="#housekeeping-and-docs-updates" title="Permanent link">&para;</a></h4>
<ul>
<li>Bump biopipen to 0.18.1</li>
<li>Mention function changes with versions in docs</li>
<li>Add apptainer in board.toml so the command can be generated in pipen-board</li>
<li>Make logo shorter in docs</li>
<li>Add docker image with <code>-full</code> tags to include all dependencies</li>
<li>Print command help message if run test failed in CI</li>
<li>Add singularity/apptainer in FAQ for "no space left" question</li>
<li>Add -w fro apptainer in docs (as we need to save pipen-board file in home directory)</li>
</ul>
<h4 id="added-080">Added-0.8.0<a class="headerlink" href="#added-080" title="Permanent link">&para;</a></h4>
<ul>
<li>Add <code>TESSA</code> process for <a href="https://pwwang.github.io/immunopipe/processes/TESSA/">tessa analysis</a></li>
<li>Add volcano plot for <code>MarkersFinder</code> and <code>ClusterMarkers</code></li>
</ul>
<h4 id="fixed">Fixed<a class="headerlink" href="#fixed" title="Permanent link">&para;</a></h4>
<ul>
<li>Fix when <code>Sample</code> is the only column in meta for <code>ImmunarchLoading</code></li>
<li>Add clear message when <code>k.weight</code> is too large for <code>IntegrateData</code> in <code>SeuratClustering</code></li>
<li>Allow <code>unique:</code> prefix for <code>on</code> in <code>SampleInfo</code></li>
<li>Fix sample order in plots for <code>SampleInfo</code></li>
<li>Remove <code>tidyseurat::</code> prefix for <code>filter</code> in scripts of <code>MetaMarkers</code>, <code>ScFGSEA</code> and <code>SeuratClusterStats</code> in case <code>tidyseurat::filter</code> is not exported when installed from <code>conda</code> (but it will make <code>dplyr::filter</code> work anyway on seurat object)</li>
</ul>
<h4 id="breaking-changes-080">Breaking changes-0.8.0<a class="headerlink" href="#breaking-changes-080" title="Permanent link">&para;</a></h4>
<ul>
<li>Redesign envs for <code>SeuratClusteringStats</code> to allow setting defaults for cases and switch identities for plots</li>
</ul>
<h3 id="070">0.7.0<a class="headerlink" href="#070" title="Permanent link">&para;</a></h3>
<h4 id="housekeeping-and-docs-updates-070">Housekeeping and docs updates-0.7.0<a class="headerlink" href="#housekeeping-and-docs-updates-070" title="Permanent link">&para;</a></h4>
<ul>
<li>Fix typos in docs/configurations</li>
<li><code>TCRClustering</code> should be <code>TCRClusteringStats</code> in Multi-case variable design section</li>
<li><code>infile</code> of <code>[SampleInfo.in]</code> should be <code>samples.txt</code> rather than <code>sample.txt</code></li>
<li>Remove unused scripts by deprecated processes</li>
<li>Bump <code>pipen-report</code> to <a href="https://github.com/pwwang/pipen-report/releases/tag/0.12.8">0.12.8</a></li>
<li>Add <code>master</code> branch and <code>master</code> tag as stable tag for docker image</li>
<li>Add pdf version of the flowchart (#4)</li>
<li>Add warning for the results in getting started tutorial</li>
<li>Bump <code>pipen-board</code> to <a href="https://github.com/pwwang/pipen-board/releases/tag/0.11.5">0.11.5</a></li>
<li>Add apptainer to the docs</li>
</ul>
<h4 id="added-070">Added-0.7.0<a class="headerlink" href="#added-070" title="Permanent link">&para;</a></h4>
<ul>
<li>Add <code>ModuleScoreCalculator</code> to calculate module scores or cell cycle scores</li>
<li>See: <a href="https://pwwang.github.io/immunopipe/processes/ModuleScoreCalculator/">https://pwwang.github.io/immunopipe/processes/ModuleScoreCalculator/</a></li>
<li>Allow <code>SampleInfo</code> to perform statistics on the sample information</li>
<li>See: <a href="https://pwwang.github.io/immunopipe/processes/SampleInfo/">https://pwwang.github.io/immunopipe/processes/SampleInfo/</a></li>
<li>Add <code>TCR_Cluster_Size</code> and <code>TCR_Cluster_Size1</code> from <code>TCRClustering</code> to metadata for further integrative analysis</li>
<li>See: <a href="https://pwwang.github.io/immunopipe/processes/TCRClusters2Seurat/">https://pwwang.github.io/immunopipe/processes/TCRClusters2Seurat/</a></li>
</ul>
<h4 id="fixed-070">Fixed-0.7.0<a class="headerlink" href="#fixed-070" title="Permanent link">&para;</a></h4>
<ul>
<li>Fix default height and width for plots in <code>SeuratClusterStats</code></li>
<li>Fix cluster order not kept after annotation using <code>hitype</code> in <code>CellTypeAnnotation</code></li>
</ul>
<h4 id="breaking-changes-070">Breaking changes-0.7.0<a class="headerlink" href="#breaking-changes-070" title="Permanent link">&para;</a></h4>
<ul>
<li>Change <code>seurat_clusters_old</code> to <code>seurat_clusters_id</code> to save old <code>seurat_clusters</code> in <code>CellTypeAnnotation</code></li>
<li>Remove <code>MarkersForClustersOfAllCells</code> and <code>TopExpressingGenesOfAllCells</code> processes</li>
<li>Rename <code>MarkersForClustersOfTCells</code> to <code>ClusterMarkers</code></li>
<li>Rename <code>TopExpressingGenesOfTCells</code> to <code>TopExpressingGenes</code></li>
<li>Rename <code>envs.exprs</code> to <code>envs.features</code> for <code>SeuratClusterStats</code></li>
<li><code>envs.exprs.genes</code> is also renamed to <code>envs.features.features</code></li>
</ul>
<h3 id="060">0.6.0<a class="headerlink" href="#060" title="Permanent link">&para;</a></h3>
<ul>
<li>⬆️ Bump biopipen to 0.16</li>
<li>📝 Add documentation</li>
<li>💚 Fix docs building in CI</li>
<li>📝 Update README with flowchart</li>
</ul>
<h3 id="051">0.5.1<a class="headerlink" href="#051" title="Permanent link">&para;</a></h3>
<ul>
<li>✨ Add <code>TopExpressingGenes</code></li>
<li>🎨 Move <code>RadarPlots</code> to <code>biopipen</code></li>
<li>⬆️ Bump biopipen to 0.15.2</li>
</ul>
<h3 id="050">0.5.0<a class="headerlink" href="#050" title="Permanent link">&para;</a></h3>
<ul>
<li>⬆️ Upgrade biopipen to 0.15.0</li>
<li>💚 Use better strategy docker image building</li>
</ul>
<h3 id="040">0.4.0<a class="headerlink" href="#040" title="Permanent link">&para;</a></h3>
<ul>
<li>⬆️ Bump biopipen to 0.6</li>
<li>⬆️ Upgrade other dependencies</li>
<li>💚 Use micromamba for docker image building</li>
<li>⬆️ Add procps-ng for vdjtools for docker  building</li>
</ul>
<h3 id="030">0.3.0<a class="headerlink" href="#030" title="Permanent link">&para;</a></h3>
<ul>
<li>💚 Use build 2 for genomeinfodbdata from bioconda (0.2.4)</li>
<li>👽️ Use config from pipen_args</li>
<li>⬆️ Pump biopipen to 0.5.3, pipen-args to 0.3.2</li>
<li>⬆️ Upgrade deps for docker</li>
<li>📝 Add flowchart in README.md</li>
<li>🐛 Fix error when --config not passed</li>
</ul>
<h3 id="024">0.2.4<a class="headerlink" href="#024" title="Permanent link">&para;</a></h3>
<ul>
<li>💚 Use lastest miniconda3 for docker build</li>
<li>💚 Use conda channel pwwang for bioconductor-genomeinfodbdata for fix (bioconda/bioconda-recipes#31349)</li>
<li>⬆️ Upgrade biopipen to 0.4.9</li>
<li>📝 Add URL to example in README</li>
</ul>
<h3 id="023">0.2.3<a class="headerlink" href="#023" title="Permanent link">&para;</a></h3>
<ul>
<li>⬆️ Upgrade biopipen to 0.4.8</li>
</ul>
<h3 id="022">0.2.2<a class="headerlink" href="#022" title="Permanent link">&para;</a></h3>
<ul>
<li>⬆️ Upgrade biopipen to 0.4.7 to fix SeuratPreparing</li>
</ul>
<h3 id="021">0.2.1<a class="headerlink" href="#021" title="Permanent link">&para;</a></h3>
<ul>
<li>🔥 Fix the bug of the wrong arguments in help page</li>
<li>⬆️ Upgrade clustcr to 1.0.2</li>
<li>📝 Fix docs for metabolic analysis</li>
</ul>
<h3 id="020">0.2.0<a class="headerlink" href="#020" title="Permanent link">&para;</a></h3>
<ul>
<li>♻️ Move in-house processes out of processes.py</li>
<li>♻️ Split up MARKERS_FINDER</li>
<li>♻️ Refactor RadarPlots</li>
<li>✨ Add an example config file</li>
<li>⚡️ Add <code>filter</code> for RadarPlots</li>
<li>📝 Update docs</li>
<li>⬆️ Upgrade deps</li>
<li>🔧 Update docker/environment.yml</li>
<li>🐛 Fix CloneHeterogeneity when only 1 row in continency table</li>
</ul>
<h3 id="011">0.1.1<a class="headerlink" href="#011" title="Permanent link">&para;</a></h3>
<ul>
<li>💚 Try fix pip in environment.yml</li>
<li>📝 Update readme for requirement checking</li>
<li>📝 Update docs to fix #1</li>
<li>📝 Update CHANGELOG</li>
<li>⬆️ Adopt biopipen 0.4.0</li>
</ul>
<h3 id="010">0.1.0<a class="headerlink" href="#010" title="Permanent link">&para;</a></h3>
<ul>
<li>🩹 Disable force-caching for some procs</li>
<li>⬆️ Upgrade datar to 0.8.*</li>
<li>✨ Add dockerfile</li>
<li>⬆️ Upgrade pipen to 0.3</li>
<li>💥 Remove gene lists from start processes</li>
<li>⬆️ Upgrade biopipen to 0.3</li>
<li>⬆️ Upgrade pipen to 0.3.5</li>
</ul>
<h3 id="007">0.0.7<a class="headerlink" href="#007" title="Permanent link">&para;</a></h3>
<ul>
<li>Add CloneHeterogeneity</li>
<li>Allow setting <code>indicator_gene</code> for <code>TCellSelection</code></li>
<li>Adopt latest datar and biopipen</li>
</ul>
<h3 id="006">0.0.6<a class="headerlink" href="#006" title="Permanent link">&para;</a></h3>
<ul>
<li>✨ Allow dimplots with clonal information</li>
</ul>
<h3 id="005">0.0.5<a class="headerlink" href="#005" title="Permanent link">&para;</a></h3>
<ul>
<li>✨ Allow more flexible dim plots</li>
</ul>
<h3 id="004">0.0.4<a class="headerlink" href="#004" title="Permanent link">&para;</a></h3>
<ul>
<li>✨ Refactor markers finder module and add meta-marker analysis</li>
</ul>
<h3 id="003">0.0.3<a class="headerlink" href="#003" title="Permanent link">&para;</a></h3>
<p>-✨ Add metabolic pathway analysis</p>
<h3 id="002">0.0.2<a class="headerlink" href="#002" title="Permanent link">&para;</a></h3>
<ul>
<li>Adopt biopipen 0.1.3</li>
</ul>
<h3 id="001">0.0.1<a class="headerlink" href="#001" title="Permanent link">&para;</a></h3>
<ul>
<li>First release</li>
</ul>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../faq/" class="btn btn-neutral float-left" title="FAQ"><span class="icon icon-circle-arrow-left"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  <div class="footer-notes">
    
    Version: <div id="mike-versions"></div>
    <div class="footer-sep"></div>
    <a href="https://github.com/pwwang/immunopipe" title="Open repository in GitHub">Github Repo</a>
    <div class="footer-sep"></div>Built with <a href="https://www.mkdocs.org/">MkDocs</a> using theme <a href="https://github.com/pwwang/mkdocs-rtd">mkdocs-rtd</a>.
  </div>
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
    
      <span><a href="../faq/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
  </span>
</div>
    <script src="../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "..";</script>
    <script src="../js/theme_extra.js"></script>
    <script src="../js/theme.js"></script>
    <script src="../js/version-select.js"></script>
      <script src="../assets/script.js"></script>
      <script src="../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
