# Preparing the input data

## Single-cell RNA-seq (scRNA-seq) data

Currently only [10X Genomics](https://www.10xgenomics.com/) data are supported.

For each sample, you need to provide a directory containing the files that are generated by [`CellRanger`](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger). Specifically, the directory should be able to be read by [`Seurat::Read_10X()`](https://satijalab.org/seurat/reference/read10x). For example, the directory should contain `matrix.mtx`, `barcodes.tsv` and `features.tsv`.
These files can also be gzipped.

You can also use the `h5` file generated by `CellRanger`.

## Single-cell TCR-seq (scTCR-seq) data

The scRNA-seq data is optional for the pipeline. However, the scTCR-seq data is required for the pipeline.

The scTCR-seq data, if available, should be paired with the scRNA-seq data. Theoratically, as long as the data can be loaded by [`immunarch::repLoad()`](https://immunarch.com/reference/repLoad.html), it should be fine. However, the pipeline is only tested with the data generated by [`CellRanger`](https://support.10xgenomics.com/single-cell-vdj/software/pipelines/latest/output/annotation). Specifically, the directory should contain `filtered_contig_annotations.csv` file, or at least the `all_contig_annotations.csv` file.

## Metadata

A metadata file is required as an input file for the pipeline. It should be a `TAB` delimited file with 3 required columns:

- `Sample`: A unique id for each sample
- `RNAData`: The directory or `h5` file for single-cell RNA data for this sample, as described above.
- `TCRData` (optional): The directory for single-cell TCR data for this sample as described above.

When `TCRData` is not provided, the pipeline will skip the processes related to scTCR-seq data (see [Routes of the pipeline](./introduction.md#only-scrna-seq-data-avaiable) for more details).
You can also add other columns to the metadata file. The columns will be added to both:

- meta data of the object loaded by `immunarch::repLoad()` (i.e. `data$meta`)
- meta data of the seurat object loaded by `Seurat::Read10X()` or `Seurat::Read10X_h5()` (i.e. `srtobj@meta.data`)

This file should be provided to `SampleInfo` process. See [`SampleInfo`](./processes/SampleInfo.md) for more details.

An example metadata file can be found [here](https://github.com/pwwang/immunopipe-example/blob/master/data/samples.txt).

You can also use [`SampleInfo`](./processes/SampleInfo.md) with `envs.save_mutated = true` and/or [`IntegratingTCR`](./processes/IntegratingTCR.md) to add columns to metadata by configuration. These columns are persisted for downstream analysis. The difference is that `SampleInfo` can only use the columns that are already in the metadata file, while `IntegratingTCR` can use the columns that are generated by the pipeline (e.g. TCR clone information).

## Other optional files

### Genes/Features to visualize for Seurat object

If you have a set of genes/features of interest, you can provide a file with those genes, one gene per line, to `SeuratClusterStats.envs.exprs.features` for:

- Ridge plots using [`Seurat::RidgePlot()`](https://satijalab.org/seurat/reference/ridgeplot),
- Violin plots using [`Seurat::VlnPlot()`](https://satijalab.org/seurat/reference/vlnplot),
- Feature plots using [`Seurat::FeaturePlot()`](https://satijalab.org/seurat/reference/featureplot),
- Dot plots using [`Seurat::DotPlot()`](https://satijalab.org/seurat/reference/dotplot), and
- Heatmaps using [`Seurat::DoHeatmap()`](https://satijalab.org/seurat/reference/doheatmap).

/// Note
The genes should exist in the RNA-seq data (i.e `features.tsv` or the `h5` file from cellranger).
///

See [`SeuratClusterStats`](./processes/SeuratClusterStats.md) for more details.

### Pathways for Gene Set Enrichment Analysis (GSEA)

If you want to perform GSEA, you need to provide a file containing the pathways. The file should be in the [GMT format][1]. You can provide the file to `ScFGSEA.envs.gmtfile`. Similarly, the genes should exist (be in the same format) in the `features.tsv` file.

See [`ScFGSEA`](./processes/ScFGSEA.md) for more details.

You can also find an example here: <https://github.com/pwwang/immunopipe-example/blob/master/data/MSigDB_Hallmark_v7.5.1.gmt>

### Cell type database for cell type annotation by `sctype` or `hitype`

If you want to perform cell type annotation, you need to provide a file containing the cell type database if you are using `sctype` or `hitype`. The database file should be fed to `CellTypeAnnotation.envs.sctype_db` if you are using `sctype`, or `CellTypeAnnotation.envs.hitype_db` if you are using `hitype`. Again, the markers in the database should exist (be in the same format) in the `features.tsv` file or the `h5` file.

See [`CellTypeAnnotation`](./processes/CellTypeAnnotation.md) for more details.

Examples can be found here: [ScTypeDB_short.xlsx](https://github.com/IanevskiAleksandr/sc-type/blob/master/ScTypeDB_short.xlsx) and [ScTypeDB_full.xlsx](https://github.com/IanevskiAleksandr/sc-type/blob/master/ScTypeDB_full.xlsx).

### Model for cell type annotation by `celltypist`

If you want to perform cell type annotation by `celltypist`, you need to provide a model file. The model file should be fed to `CellTypeAnnotation.envs.celltypist_args.model`. The information of models can be found [here](https://celltypist.cog.sanger.ac.uk/models/models.json). Download the one you want to use and provide the path to the file.

### Metabolic pathway for Metabolic Landscape Analysis

Similarly, if you want to perform metabolic landscape analysis, you need to provide a file containing the metabolic pathways. The file should be in the [GMT format][1]. You can provide the file to `ScMetabolicLandscape.envs.gmtfile`. This file can also be used for GSEA. A pathway file for KEGG metabolism is provided [here][2].

See [`ScrnaMetabolicLandscape`](./processes/ScrnaMetabolicLandscape.md) for more details.

### Reference for Seurat mapping if you want to perform supervised clustering

If you want to perform supervised clustering, you need to provide a reference for [`SeuratMap2Ref`](./processes/SeuratMap2Ref.md). The reference should be a `Seurat` object in `RDS` or `h5seurat` file. You can provide the reference to `SeuratMap2Ref.envs.ref`.

See [`SeuratMap2Ref`](./processes/SeuratMap2Ref.md) for more details.

[1]: https://docs.gsea-msigdb.org/#GSEA/Data_Formats/#gmt-gene-matrix-transposed-file-format-gmt
[2]: https://github.com/pwwang/biopipen/blob/master/tests/data/scrna_metabolic/KEGG_metabolism.gmt
