<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      <link rel="shortcut icon" href="../../img/favicon.ico" />
    <title>PseudoBulkDEG - immunopipe</title>
    <link rel="stylesheet" href="../../css/theme.css" />
    <link rel="stylesheet" href="../../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
        <link href="../../assets/style.css" rel="stylesheet" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "PseudoBulkDEG";
        var mkdocs_page_input_path = "processes/PseudoBulkDEG.md";
        var mkdocs_page_url = null;
      </script>
    
    <!--[if lt IE 9]>
      <script src="../../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.configure({ ignoreUnescapedHTML: true }); hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="../..">
          <img src="../../rtd-logo.png" class="logo" alt="Logo"/>
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="nav-scrollable-wrapper">
      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../..">Home</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../introduction/">Introduction</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../installation/">Installation</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../getting-started/">Geting started</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../gallery/">Gallery</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../running/">Running the pipeline</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../preparing-input/">Preparing the input</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../configurations/">Configurations</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">Processes</span></p>
              <ul class="current">
                  <li class="toctree-l1"><a class="reference internal" href="../SampleInfo/">SampleInfo</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../SeuratPreparing/">SeuratPreparing</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../SeuratClusteringOfAllCells/">SeuratClusteringOfAllCells</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ClusterMarkersOfAllCells/">ClusterMarkersOfAllCells</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../TopExpressingGenesOfAllCells/">TopExpressingGenesOfAllCells</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../TCellSelection/">TCellSelection</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ModuleScoreCalculator/">ModuleScoreCalculator</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../SeuratClustering/">SeuratClustering</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../CellTypeAnnotation/">CellTypeAnnotation</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../SeuratMap2Ref/">SeuratMap2Ref</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ClusterMarkers/">ClusterMarkers</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../TopExpressingGenes/">TopExpressingGenes</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../SeuratSubClustering/">SeuratSubClustering</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ScRepLoading/">ScRepLoading</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ScRepCombiningExpression/">ScRepCombiningExpression</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../TCRClustering/">TCRClustering</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../TESSA/">TESSA</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../SeuratClusterStats/">SeuratClusterStats</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ScFGSEA/">ScFGSEA</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../MarkersFinder/">MarkersFinder</a>
                  </li>
                  <li class="toctree-l1 current"><a class="reference internal current" href="./">PseudoBulkDEG</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#input">Input</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#output">Output</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#environment-variables">Environment Variables</a>
    </li>
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ClonalStats/">ClonalStats</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../CDR3AAPhyschem/">CDR3AAPhyschem</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../CellCellCommunication/">CellCellCommunication</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../CellCellCommunicationPlots/">CellCellCommunicationPlots</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Metabolic Landscape Analysis</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../ScrnaMetabolicLandscape/">Introduction & Group Arguments</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../MetabolicInput/">MetabolicInput</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../MetabolicExprImputation/">MetabolicExprImputation</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../MetabolicFeatures/">MetabolicFeatures</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../MetabolicPathwayActivity/">MetabolicPathwayActivity</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../MetabolicPathwayHeterogeneity/">MetabolicPathwayHeterogeneity</a>
                  </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../faq/">FAQ</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../CHANGELOG/">Change log</a>
                </li>
              </ul>
      </div>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../..">immunopipe</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../.." class="icon icon-home" aria-label="Docs"></a></li>
          <li class="breadcrumb-item">Processes</li>
      <li class="breadcrumb-item active">PseudoBulkDEG</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/pwwang/immunopipe/blob/dev/docs/processes/PseudoBulkDEG.md">Edit on pwwang/immunopipe</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h2 id="pseudobulkdeg">PseudoBulkDEG<a class="headerlink" href="#pseudobulkdeg" title="Permanent link">&para;</a></h2>
<p>Pseduo-bulk differential gene expression analysis</p>
<p>This process performs differential gene expression analysis, instead of
on single-cell level, on the pseudo-bulk data, aggregated from the single-cell data.<br /></p>
<h3 id="input">Input<a class="headerlink" href="#input" title="Permanent link">&para;</a></h3>
<ul>
<li><code>sobjfile</code>:
    The seurat object file in RDS or qs/qs2 format.<br /></li>
</ul>
<h3 id="output">Output<a class="headerlink" href="#output" title="Permanent link">&para;</a></h3>
<ul>
<li><code>outdir</code>: <em>Default: <code>{{in.sobjfile | stem}}.pseudobulk_deg</code></em>. <br />
    The output containing the results of the differential gene expression
    analysis.<br /></li>
</ul>
<h3 id="environment-variables">Environment Variables<a class="headerlink" href="#environment-variables" title="Permanent link">&para;</a></h3>
<ul>
<li><code>mutaters</code> <em>(<code>type=json</code>)</em>: <em>Default: <code>{}</code></em>. <br />
    Mutaters to mutate the metadata of the
    seurat object. Keys are the new column names and values are the
    expressions to mutate the columns. These new columns can be
    used to define your cases.<br />
    You can also use the clone selectors to select the TCR clones/clusters.<br />
    See <a href="https://pwwang.github.io/scplotter/reference/clone_selectors.html">https://pwwang.github.io/scplotter/reference/clone_selectors.html</a>.<br /></li>
<li><code>each</code>:
    The column name in metadata to separate the cells into different cases.<br />
    When specified, the case will be expanded to multiple cases for
    each value in the column.<br /></li>
<li><code>subset</code>:
    An expression in string to subset the cells.<br /></li>
<li><code>aggregate_by</code>:
    The column names in metadata to aggregate the cells.<br /></li>
<li><code>layer</code>: <em>Default: <code>counts</code></em>. <br />
    The layer to pull and aggregate the data.<br /></li>
<li><code>assay</code>: <em>Default: <code>RNA</code></em>. <br />
    The assay to pull and aggregate the data.<br /></li>
<li><code>error</code> <em>(<code>flag</code>)</em>: <em>Default: <code>False</code></em>. <br />
    Error out if no/not enough markers are found or no pathways are enriched.<br />
    If <code>False</code>, empty results will be returned.<br /></li>
<li><code>group_by</code>:
    The column name in metadata to group the cells.<br /></li>
<li><code>ident_1</code>:
    The first identity to compare.<br /></li>
<li><code>ident_2</code>:
    The second identity to compare.<br />
    If not specified, the rest of the identities will be compared with <code>ident_1</code>.<br /></li>
<li><code>paired_by</code>:
    The column name in metadata to mark the paired samples.<br />
    For example, subject. If specified, the paired test will be performed.<br /></li>
<li><code>dbs</code> <em>(<code>list</code>)</em>: <em>Default: <code>['KEGG_2021_Human', 'MSigDB_Hallmark_2020']</code></em>. <br />
    The databases to use for enrichment analysis.<br />
    The databases are passed to <code>biopipen.utils::Enrichr()</code> to do the
    enrichment analysis. The default databases are <code>KEGG_2021_Human</code> and
    <code>MSigDB_Hallmark_2020</code>.<br />
    See <a href="https://maayanlab.cloud/Enrichr/#libraries">https://maayanlab.cloud/Enrichr/#libraries</a> for the available
    libraries.<br /></li>
<li><code>sigmarkers</code>: <em>Default: <code>p_val_adj &lt; 0.05</code></em>. <br />
    An expression passed to <code>dplyr::filter()</code> to filter the
    significant markers for enrichment analysis.<br />
    The default is <code>p_val_adj &lt; 0.05</code>.<br />
    If <code>tool = 'DESeq2'</code>, the variables that can be used for filtering
    are: <code>baseMean</code>, <code>log2FC</code>, <code>lfcSE</code>, <code>stat</code>, <code>p_val</code>, <code>p_val_adj</code>.<br />
    If <code>tool = 'edgeR'</code>, the variables that can be used for filtering
    are: <code>logCPM</code>, <code>log2FC</code>, <code>LR</code>, <code>p_val</code>, <code>p_val_adj</code>.<br /></li>
<li><code>enrich_style</code> <em>(<code>choice</code>)</em>: <em>Default: <code>enrichr</code></em>. <br />
    The style of the enrichment analysis.<br /><ul>
<li><code>enrichr</code>:
    Use <code>enrichr</code>-style for the enrichment analysis.<br /></li>
<li><code>clusterProfiler</code>:
    Use <code>clusterProfiler</code>-style for the enrichment analysis.<br /></li>
</ul>
</li>
<li><code>allmarker_plots_defaults</code> <em>(<code>ns</code>)</em>:
    Default options for the plots for all markers when <code>ident-1</code> is not specified.<br /><ul>
<li><code>plot_type</code>:
    The type of the plot.<br />
    See <a href="https://pwwang.github.io/scplotter/reference/FeatureStatPlot.html">https://pwwang.github.io/scplotter/reference/FeatureStatPlot.html</a>.<br />
    Available types are <code>violin</code>, <code>box</code>, <code>bar</code>, <code>ridge</code>, <code>dim</code>, <code>heatmap</code> and <code>dot</code>.<br /></li>
<li><code>more_formats</code> <em>(<code>type=list</code>)</em>: <em>Default: <code>[]</code></em>. <br />
    The extra formats to save the plot in.<br /></li>
<li><code>save_code</code> <em>(<code>flag</code>)</em>: <em>Default: <code>False</code></em>. <br />
    Whether to save the code to generate the plot.<br /></li>
<li><code>devpars</code> <em>(<code>ns</code>)</em>:
    The device parameters for the plots.<br /><ul>
<li><code>res</code> <em>(<code>type=int</code>)</em>: <em>Default: <code>100</code></em>. <br />
    The resolution of the plots.<br /></li>
<li><code>height</code> <em>(<code>type=int</code>)</em>:
    The height of the plots.<br /></li>
<li><code>width</code> <em>(<code>type=int</code>)</em>:
    The width of the plots.<br /></li>
</ul>
</li>
<li><code>order_by</code>: <em>Default: <code>desc(abs(log2FC))</code></em>. <br />
    an expression to order the markers, passed by <code>dplyr::arrange()</code>.<br /></li>
<li><code>genes</code>: <em>Default: <code>10</code></em>. <br />
    The number of top genes to show or an expression passed to <code>dplyr::filter()</code> to filter the genes.<br /></li>
<li><code>&lt;more&gt;</code>:
    Other arguments passed to <a href="https://pwwang.github.io/scplotter/reference/FeatureStatPlot.html"><code>scplotter::FeatureStatPlot()</code></a>.<br /></li>
</ul>
</li>
<li><code>allmarker_plots</code> <em>(<code>type=json</code>)</em>: <em>Default: <code>{}</code></em>. <br />
    All marker plot cases.<br />
    The keys are the names of the cases and the values are the dicts inherited from <code>allmarker_plots_defaults</code>.<br /></li>
<li><code>allenrich_plots_defaults</code> <em>(<code>ns</code>)</em>:
    Default options for the plots to generate for the enrichment analysis.<br /><ul>
<li><code>plot_type</code>: <em>Default: <code>heatmap</code></em>. <br />
    The type of the plot.<br /></li>
<li><code>devpars</code> <em>(<code>ns</code>)</em>:
    The device parameters for the plots.<br /><ul>
<li><code>res</code> <em>(<code>type=int</code>)</em>: <em>Default: <code>100</code></em>. <br />
    The resolution of the plots.<br /></li>
<li><code>height</code> <em>(<code>type=int</code>)</em>:
    The height of the plots.<br /></li>
<li><code>width</code> <em>(<code>type=int</code>)</em>:
    The width of the plots.<br /></li>
</ul>
</li>
<li><code>&lt;more&gt;</code>:
    See <a href="https://pwwang.github.io/scplotter/reference/EnrichmentPlot.html">https://pwwang.github.io/scplotter/reference/EnrichmentPlot.html</a>.<br /></li>
</ul>
</li>
<li><code>allenrich_plots</code> <em>(<code>type=json</code>)</em>: <em>Default: <code>{}</code></em>. <br />
    Cases of the plots to generate for the enrichment analysis.<br />
    The keys are the names of the cases and the values are the dicts inherited from <code>allenrich_plots_defaults</code>.<br />
    The cases under <code>envs.cases</code> can inherit this options.<br /></li>
<li><code>marker_plots_defaults</code> <em>(<code>ns</code>)</em>:
    Default options for the plots to generate for the markers.<br /><ul>
<li><code>plot_type</code>:
    The type of the plot.<br />
    See <a href="https://pwwang.github.io/scplotter/reference/FeatureStatPlot.html">https://pwwang.github.io/scplotter/reference/FeatureStatPlot.html</a>.<br />
    Available types are <code>violin</code>, <code>box</code>, <code>bar</code>, <code>ridge</code>, <code>dim</code>, <code>heatmap</code> and <code>dot</code>.<br />
    There are two additional types available - <code>volcano_pct</code> and <code>volcano_log2fc</code>.<br /></li>
<li><code>more_formats</code> <em>(<code>type=list</code>)</em>: <em>Default: <code>[]</code></em>. <br />
    The extra formats to save the plot in.<br /></li>
<li><code>save_code</code> <em>(<code>flag</code>)</em>: <em>Default: <code>False</code></em>. <br />
    Whether to save the code to generate the plot.<br /></li>
<li><code>devpars</code> <em>(<code>ns</code>)</em>:
    The device parameters for the plots.<br /><ul>
<li><code>res</code> <em>(<code>type=int</code>)</em>: <em>Default: <code>100</code></em>. <br />
    The resolution of the plots.<br /></li>
<li><code>height</code> <em>(<code>type=int</code>)</em>:
    The height of the plots.<br /></li>
<li><code>width</code> <em>(<code>type=int</code>)</em>:
    The width of the plots.<br /></li>
</ul>
</li>
<li><code>order_by</code>: <em>Default: <code>desc(abs(log2FC))</code></em>. <br />
    an expression to order the markers, passed by <code>dplyr::arrange()</code>.<br /></li>
<li><code>genes</code>: <em>Default: <code>10</code></em>. <br />
    The number of top genes to show or an expression passed to <code>dplyr::filter()</code> to filter the genes.<br /></li>
<li><code>&lt;more&gt;</code>:
    Other arguments passed to <a href="https://pwwang.github.io/scplotter/reference/FeatureStatPlot.html"><code>scplotter::FeatureStatPlot()</code></a>.<br />
    If <code>plot_type</code> is <code>volcano_pct</code> or <code>volcano_log2fc</code>, they will be passed to
    <a href="https://pwwang.github.io/plotthis/reference/VolcanoPlot.html"><code>scplotter::VolcanoPlot()</code></a>.<br /></li>
</ul>
</li>
<li><code>marker_plots</code> <em>(<code>type=json</code>)</em>: <em>Default: <code>{'Volcano Plot': Diot({'plot_type': 'volcano'})}</code></em>. <br />
    Cases of the plots to generate for the markers.<br />
    Plot cases. The keys are the names of the cases and the values are the dicts inherited from <code>marker_plots_defaults</code>.<br />
    The cases under <code>envs.cases</code> can inherit this options.<br /></li>
<li><code>enrich_plots_defaults</code> <em>(<code>ns</code>)</em>:
    Default options for the plots to generate for the enrichment analysis.<br /><ul>
<li><code>plot_type</code>:
    The type of the plot.<br />
    See <a href="https://pwwang.github.io/scplotter/reference/EnrichmentPlot.html">https://pwwang.github.io/scplotter/reference/EnrichmentPlot.html</a>.<br />
    Available types are <code>bar</code>, <code>dot</code>, <code>lollipop</code>, <code>network</code>, <code>enrichmap</code> and <code>wordcloud</code>.<br /></li>
<li><code>more_formats</code> <em>(<code>type=list</code>)</em>: <em>Default: <code>[]</code></em>. <br />
    The extra formats to save the plot in.<br /></li>
<li><code>save_code</code> <em>(<code>flag</code>)</em>: <em>Default: <code>False</code></em>. <br />
    Whether to save the code to generate the plot.<br /></li>
<li><code>devpars</code> <em>(<code>ns</code>)</em>:
    The device parameters for the plots.<br /><ul>
<li><code>res</code> <em>(<code>type=int</code>)</em>: <em>Default: <code>100</code></em>. <br />
    The resolution of the plots.<br /></li>
<li><code>height</code> <em>(<code>type=int</code>)</em>:
    The height of the plots.<br /></li>
<li><code>width</code> <em>(<code>type=int</code>)</em>:
    The width of the plots.<br /></li>
</ul>
</li>
<li><code>&lt;more&gt;</code>:
    See <a href="https://pwwang.github.io/scplotter/reference/EnrichmentPlot.htmll">https://pwwang.github.io/scplotter/reference/EnrichmentPlot.htmll</a>.<br /></li>
</ul>
</li>
<li><code>enrich_plots</code> <em>(<code>type=json</code>)</em>: <em>Default: <code>{'Bar Plot': Diot({'plot_type': 'bar', 'ncol': 1, 'top_term': 10})}</code></em>. <br />
    Cases of the plots to generate for the enrichment analysis.<br />
    The keys are the names of the cases and the values are the dicts inherited from <code>enrich_plots_defaults</code>.<br />
    The cases under <code>envs.cases</code> can inherit this options.<br /></li>
<li><code>overlaps_defaults</code> <em>(<code>ns</code>)</em>:
    Default options for investigating the overlapping of significant markers between different cases or comparisons.<br />
    This means either <code>ident-1</code> should be empty, so that they can be expanded to multiple comparisons.<br /><ul>
<li><code>sigmarkers</code>:
    The expression to filter the significant markers for each case.<br />
    If not provided, <code>envs.sigmarkers</code> will be used.<br /></li>
<li><code>plot_type</code> <em>(<code>choice</code>)</em>: <em>Default: <code>venn</code></em>. <br />
    The type of the plot to generate for the overlaps.<br /><ul>
<li><code>venn</code>:
    Use <code>plotthis::VennDiagram()</code>.<br /></li>
<li><code>upset</code>:
    Use <code>plotthis::UpsetPlot()</code>.<br /></li>
</ul>
</li>
<li><code>more_formats</code> <em>(<code>type=list</code>)</em>: <em>Default: <code>[]</code></em>. <br />
    The extra formats to save the plot in.<br /></li>
<li><code>save_code</code> <em>(<code>flag</code>)</em>: <em>Default: <code>False</code></em>. <br />
    Whether to save the code to generate the plot.<br /></li>
<li><code>devpars</code> <em>(<code>ns</code>)</em>:
    The device parameters for the plots.<br /><ul>
<li><code>res</code> <em>(<code>type=int</code>)</em>: <em>Default: <code>100</code></em>. <br />
    The resolution of the plots.<br /></li>
<li><code>height</code> <em>(<code>type=int</code>)</em>:
    The height of the plots.<br /></li>
<li><code>width</code> <em>(<code>type=int</code>)</em>:
    The width of the plots.<br /></li>
</ul>
</li>
<li><code>&lt;more&gt;</code>:
    More arguments pased to <code>plotthis::VennDiagram()</code>
    (<a href="https://pwwang.github.io/plotthis/reference/venndiagram1.html">https://pwwang.github.io/plotthis/reference/venndiagram1.html</a>)
    or <code>plotthis::UpsetPlot()</code>
    (<a href="https://pwwang.github.io/plotthis/reference/upsetplot1.html">https://pwwang.github.io/plotthis/reference/upsetplot1.html</a>)</li>
</ul>
</li>
<li><code>overlaps</code> <em>(<code>type=json</code>)</em>: <em>Default: <code>{}</code></em>. <br />
    Cases for investigating the overlapping of significant markers between different cases or comparisons.<br />
    The keys are the names of the cases and the values are the dicts inherited from <code>overlaps_defaults</code>.<br />
    There are two situations that we can perform overlaps:<br /><ol>
<li>If <code>ident-1</code> is not specified, the overlaps can be performed between different comparisons.<br /></li>
<li>If <code>each</code> is specified, the overlaps can be performed between different cases, where in each case, <code>ident-1</code> must be specified.<br /></li>
</ol>
</li>
<li><code>tool</code> <em>(<code>choice</code>)</em>: <em>Default: <code>DESeq2</code></em>. <br />
    The method to use for the differential expression analysis.<br /><ul>
<li><code>DESeq2</code>:
    Use DESeq2 for the analysis.<br /></li>
<li><code>edgeR</code>:
    Use edgeR for the analysis.<br /></li>
</ul>
</li>
<li><code>plots_defaults</code> <em>(<code>ns</code>)</em>:
    The default parameters for the plots.<br /><ul>
<li><code>&lt;more&gt;</code>:
    Parameters passed to <code>biopipen.utils::VizBulkDEGs()</code>.<br />
    See: <a href="https://pwwang.github.io/biopipen.utils.R/reference/VizBulkDEGs.html">https://pwwang.github.io/biopipen.utils.R/reference/VizBulkDEGs.html</a></li>
</ul>
</li>
<li><code>plots</code> <em>(<code>type=json</code>)</em>:
    The parameters for the plots.<br />
    The keys are the names of the plots and the values are the parameters
    for the plots. The parameters will override the defaults in <code>plots_defaults</code>.<br />
    If not specified, no plots will be generated.<br /></li>
<li><code>cases</code> <em>(<code>type=json</code>)</em>: <em>Default: <code>{}</code></em>. <br />
    The cases for the analysis.<br />
    The keys are the names of the cases and the values are the arguments for
    the analysis. The arguments include the ones inherited from <code>envs</code>.<br />
    If no cases are specified, a default case will be added with
    the name <code>DEG Analysis</code> and the default values specified above.<br /></li>
</ul>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../MarkersFinder/" class="btn btn-neutral float-left" title="MarkersFinder"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../ClonalStats/" class="btn btn-neutral float-right" title="ClonalStats">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  <div class="footer-notes">
    
    Version: <div id="mike-versions"></div>
    <div class="footer-sep"></div>
    <a href="https://github.com/pwwang/immunopipe" title="Open repository in GitHub">Github Repo</a>
    <div class="footer-sep"></div>Built with <a href="https://www.mkdocs.org/">MkDocs</a> using theme <a href="https://github.com/pwwang/mkdocs-rtd">mkdocs-rtd</a>.
  </div>
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
    
      <span><a href="../MarkersFinder/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../ClonalStats/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script src="../../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "../..";</script>
    <script src="../../js/theme_extra.js"></script>
    <script src="../../js/theme.js"></script>
    <script src="../../js/version-select.js"></script>
      <script src="../../assets/script.js"></script>
      <script src="../../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
